// json-joy/lib/json-crdt-patch/PatchBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PatchBuilder = void 0;
var tslib_1 = require("tslib");
var ArrInsOp_1 = require("./operations/ArrInsOp");
var ArrOp_1 = require("./operations/ArrOp");
var BinInsOp_1 = require("./operations/BinInsOp");
var BinOp_1 = require("./operations/BinOp");
var ConstOp_1 = require("./operations/ConstOp");
var DelOp_1 = require("./operations/DelOp");
var clock_1 = require("./clock");
var isUint8Array_1 = require("../util/buffers/isUint8Array");
var NoopOp_1 = require("./operations/NoopOp");
var ObjOp_1 = require("./operations/ObjOp");
var ObjSetOp_1 = require("./operations/ObjSetOp");
var Patch_1 = require("./Patch");
var StrInsOp_1 = require("./operations/StrInsOp");
var StrOp_1 = require("./operations/StrOp");
var ValOp_1 = require("./operations/ValOp");
var ValSetOp_1 = require("./operations/ValSetOp");
var TupOp_1 = require("./operations/TupOp");
var constants_1 = require("./constants");
var Tuple_1 = require("./builder/Tuple");
var Konst_1 = require("./builder/Konst");
var DelayedValueBuilder_1 = require("./builder/DelayedValueBuilder");
var maybeConst = function (x) {
    switch (typeof x) {
        case 'number':
        case 'boolean':
            return true;
        default:
            return x === null;
    }
};
var PatchBuilder = (function () {
    function PatchBuilder(clock) {
        this.clock = clock;
        this.patch = new Patch_1.Patch();
    }
    PatchBuilder.prototype.nextTime = function () {
        return this.patch.nextTime() || this.clock.time;
    };
    PatchBuilder.prototype.flush = function () {
        var patch = this.patch;
        this.patch = new Patch_1.Patch();
        return patch;
    };
    PatchBuilder.prototype.obj = function () {
        this.pad();
        var id = this.clock.tick(1);
        this.patch.ops.push(new ObjOp_1.ObjOp(id));
        return id;
    };
    PatchBuilder.prototype.arr = function () {
        this.pad();
        var id = this.clock.tick(1);
        this.patch.ops.push(new ArrOp_1.ArrOp(id));
        return id;
    };
    PatchBuilder.prototype.tup = function () {
        this.pad();
        var id = this.clock.tick(1);
        this.patch.ops.push(new TupOp_1.TupOp(id));
        return id;
    };
    PatchBuilder.prototype.str = function () {
        this.pad();
        var id = this.clock.tick(1);
        this.patch.ops.push(new StrOp_1.StrOp(id));
        return id;
    };
    PatchBuilder.prototype.bin = function () {
        this.pad();
        var id = this.clock.tick(1);
        this.patch.ops.push(new BinOp_1.BinOp(id));
        return id;
    };
    PatchBuilder.prototype.const = function (value) {
        this.pad();
        var id = this.clock.tick(1);
        this.patch.ops.push(new ConstOp_1.ConstOp(id, value));
        return id;
    };
    PatchBuilder.prototype.val = function (val) {
        this.pad();
        var id = this.clock.tick(1);
        this.patch.ops.push(new ValOp_1.ValOp(id, val));
        return id;
    };
    PatchBuilder.prototype.root = function (val) {
        this.pad();
        var id = this.clock.tick(1);
        this.patch.ops.push(new ValSetOp_1.ValSetOp(id, constants_1.ORIGIN, val));
        return id;
    };
    PatchBuilder.prototype.setKeys = function (obj, data) {
        this.pad();
        if (!data.length)
            throw new Error('EMPTY_TUPLES');
        var id = this.clock.tick(0);
        var op = new ObjSetOp_1.ObjSetOp(id, obj, data);
        var span = op.span();
        if (span > 1)
            this.clock.tick(span - 1);
        this.patch.ops.push(op);
        return id;
    };
    PatchBuilder.prototype.setVal = function (obj, val) {
        this.pad();
        var id = this.clock.tick(0);
        var op = new ValSetOp_1.ValSetOp(id, obj, val);
        this.patch.ops.push(op);
        return id;
    };
    PatchBuilder.prototype.insStr = function (obj, ref, data) {
        this.pad();
        if (!data.length)
            throw new Error('EMPTY_STRING');
        var id = this.clock.tick(1);
        var op = new StrInsOp_1.StrInsOp(id, obj, ref, data);
        var span = op.span();
        if (span > 1)
            this.clock.tick(span - 1);
        this.patch.ops.push(op);
        return id;
    };
    PatchBuilder.prototype.insBin = function (obj, ref, data) {
        this.pad();
        if (!data.length)
            throw new Error('EMPTY_BINARY');
        var id = this.clock.tick(1);
        var op = new BinInsOp_1.BinInsOp(id, obj, ref, data);
        var span = op.span();
        if (span > 1)
            this.clock.tick(span - 1);
        this.patch.ops.push(op);
        return id;
    };
    PatchBuilder.prototype.insArr = function (arr, ref, data) {
        this.pad();
        var id = this.clock.tick(1);
        var op = new ArrInsOp_1.ArrInsOp(id, arr, ref, data);
        var span = op.span();
        if (span > 1)
            this.clock.tick(span - 1);
        this.patch.ops.push(op);
        return id;
    };
    PatchBuilder.prototype.del = function (obj, what) {
        this.pad();
        var id = this.clock.tick(0);
        this.patch.ops.push(new DelOp_1.DelOp(id, obj, what));
        return id;
    };
    PatchBuilder.prototype.noop = function (span) {
        this.pad();
        var id = this.clock.tick(span);
        this.patch.ops.push(new NoopOp_1.NoopOp(id, span));
        return id;
    };
    PatchBuilder.prototype.jsonObj = function (json) {
        var e_1, _a;
        var obj = this.obj();
        var keys = Object.keys(json);
        if (keys.length) {
            var tuples = [];
            try {
                for (var keys_1 = tslib_1.__values(keys), keys_1_1 = keys_1.next(); !keys_1_1.done; keys_1_1 = keys_1.next()) {
                    var k = keys_1_1.value;
                    var value = json[k];
                    var valueId = value instanceof clock_1.Timestamp ? value : maybeConst(value) ? this.const(value) : this.json(value);
                    tuples.push([k, valueId]);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (keys_1_1 && !keys_1_1.done && (_a = keys_1.return)) _a.call(keys_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            this.setKeys(obj, tuples);
        }
        return obj;
    };
    PatchBuilder.prototype.jsonArr = function (json) {
        var e_2, _a;
        var arr = this.arr();
        if (json.length) {
            var values = [];
            try {
                for (var json_1 = tslib_1.__values(json), json_1_1 = json_1.next(); !json_1_1.done; json_1_1 = json_1.next()) {
                    var el = json_1_1.value;
                    values.push(this.json(el));
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (json_1_1 && !json_1_1.done && (_a = json_1.return)) _a.call(json_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
            this.insArr(arr, arr, values);
        }
        return arr;
    };
    PatchBuilder.prototype.jsonStr = function (json) {
        var str = this.str();
        if (json)
            this.insStr(str, str, json);
        return str;
    };
    PatchBuilder.prototype.jsonBin = function (json) {
        var bin = this.bin();
        if (json.length)
            this.insBin(bin, bin, json);
        return bin;
    };
    PatchBuilder.prototype.jsonVal = function (json) {
        var val = this.const(json);
        return this.val(val);
    };
    PatchBuilder.prototype.jsonTup = function (slots) {
        var tup = this.tup();
        var length = slots.length;
        if (length) {
            var writes = [];
            for (var i = 0; i < length; i++)
                writes.push([i, this.constOrJson(slots[i])]);
            this.setKeys(tup, writes);
        }
        return tup;
    };
    PatchBuilder.prototype.json = function (json) {
        if (json instanceof clock_1.Timestamp)
            return json;
        if (json === undefined)
            return this.const(json);
        if (json instanceof Array)
            return this.jsonArr(json);
        if ((0, isUint8Array_1.isUint8Array)(json))
            return this.jsonBin(json);
        if (json instanceof Tuple_1.Tuple)
            return this.jsonTup(json.slots);
        if (json instanceof Konst_1.Konst)
            return this.const(json.val);
        if (json instanceof DelayedValueBuilder_1.DelayedValueBuilder)
            return json.build(this);
        switch (typeof json) {
            case 'object':
                return json === null ? this.jsonVal(json) : this.jsonObj(json);
            case 'string':
                return this.jsonStr(json);
            case 'number':
            case 'boolean':
                return this.jsonVal(json);
        }
        throw new Error('INVALID_JSON');
    };
    PatchBuilder.prototype.constOrJson = function (value) {
        if (value instanceof clock_1.Timestamp)
            return value;
        return maybeConst(value) ? this.const(value) : this.json(value);
    };
    PatchBuilder.prototype.maybeConst = function (value) {
        return value instanceof clock_1.Timestamp ? value : this.const(value);
    };
    PatchBuilder.prototype.pad = function () {
        var nextTime = this.patch.nextTime();
        if (!nextTime)
            return;
        var drift = this.clock.time - nextTime;
        if (drift > 0) {
            var id = (0, clock_1.ts)(this.clock.sid, nextTime);
            var padding = new NoopOp_1.NoopOp(id, drift);
            this.patch.ops.push(padding);
        }
    };
    return PatchBuilder;
}());
exports.PatchBuilder = PatchBuilder;
