// json-joy/lib/json-pack/json/JsonDecoder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonDecoder = void 0;
var decodeUtf8_1 = require("../../util/buffers/utf8/decodeUtf8");
var Reader_1 = require("../../util/buffers/Reader");
var fromBase64Bin_1 = require("../../util/base64/fromBase64Bin");
var REGEX_REPLACE_ESCAPED_CHARS = /\\(b|f|n|r|t|"|\/|\\)/g;
var escapedCharReplacer = function (char) {
    switch (char) {
        case '\\b':
            return '\b';
        case '\\f':
            return '\f';
        case '\\n':
            return '\n';
        case '\\r':
            return '\r';
        case '\\t':
            return '\t';
        case '\\"':
            return '"';
        case '\\/':
            return '/';
        case '\\\\':
            return '\\';
    }
    return char;
};
var hasBinaryPrefix = function (u8, x) {
    return u8[x] === 0x64 &&
        u8[x + 1] === 0x61 &&
        u8[x + 2] === 0x74 &&
        u8[x + 3] === 0x61 &&
        u8[x + 4] === 0x3a &&
        u8[x + 5] === 0x61 &&
        u8[x + 6] === 0x70 &&
        u8[x + 7] === 0x70 &&
        u8[x + 8] === 0x6c &&
        u8[x + 9] === 0x69 &&
        u8[x + 10] === 0x63 &&
        u8[x + 11] === 0x61 &&
        u8[x + 12] === 0x74 &&
        u8[x + 13] === 0x69 &&
        u8[x + 14] === 0x6f &&
        u8[x + 15] === 0x6e &&
        u8[x + 16] === 0x2f &&
        u8[x + 17] === 0x6f &&
        u8[x + 18] === 0x63 &&
        u8[x + 19] === 0x74 &&
        u8[x + 20] === 0x65 &&
        u8[x + 21] === 0x74 &&
        u8[x + 22] === 0x2d &&
        u8[x + 23] === 0x73 &&
        u8[x + 24] === 0x74 &&
        u8[x + 25] === 0x72 &&
        u8[x + 26] === 0x65 &&
        u8[x + 27] === 0x61 &&
        u8[x + 28] === 0x6d &&
        u8[x + 29] === 0x3b &&
        u8[x + 30] === 0x62 &&
        u8[x + 31] === 0x61 &&
        u8[x + 32] === 0x73 &&
        u8[x + 33] === 0x65 &&
        u8[x + 34] === 0x36 &&
        u8[x + 35] === 0x34 &&
        u8[x + 36] === 0x2c;
};
var findEndingQuote = function (uint8, x) {
    var len = uint8.length;
    var char = uint8[x];
    var prev = 0;
    while (x < len) {
        if (char === 34 && prev !== 92)
            break;
        if (char === 92 && prev === 92)
            prev = 0;
        else
            prev = char;
        char = uint8[++x];
    }
    if (x === len)
        throw new Error('Invalid JSON');
    return x;
};
var fromCharCode = String.fromCharCode;
var readShortUtf8StrAndUnescape = function (reader) {
    var buf = reader.uint8;
    var len = buf.length;
    var points = [];
    var x = reader.x;
    var prev = 0;
    while (x < len) {
        var code = buf[x++];
        if ((code & 0x80) === 0) {
            if (prev === 92) {
                switch (code) {
                    case 98:
                        code = 8;
                        break;
                    case 102:
                        code = 12;
                        break;
                    case 110:
                        code = 10;
                        break;
                    case 114:
                        code = 13;
                        break;
                    case 116:
                        code = 9;
                        break;
                    case 34:
                        code = 34;
                        break;
                    case 47:
                        code = 47;
                        break;
                    case 92:
                        code = 92;
                        break;
                    default:
                        throw new Error('Invalid JSON');
                }
                prev = 0;
            }
            else {
                if (code === 34)
                    break;
                prev = code;
                if (prev === 92)
                    continue;
            }
        }
        else {
            var octet2 = buf[x++] & 0x3f;
            if ((code & 0xe0) === 0xc0) {
                code = ((code & 0x1f) << 6) | octet2;
            }
            else {
                var octet3 = buf[x++] & 0x3f;
                if ((code & 0xf0) === 0xe0) {
                    code = ((code & 0x1f) << 12) | (octet2 << 6) | octet3;
                }
                else {
                    if ((code & 0xf8) === 0xf0) {
                        var octet4 = buf[x++] & 0x3f;
                        var unit = ((code & 0x07) << 0x12) | (octet2 << 0x0c) | (octet3 << 0x06) | octet4;
                        if (unit > 0xffff) {
                            unit -= 0x10000;
                            var unit0 = ((unit >>> 10) & 0x3ff) | 0xd800;
                            unit = 0xdc00 | (unit & 0x3ff);
                            points.push(unit0);
                            code = unit;
                        }
                        else {
                            code = unit;
                        }
                    }
                }
            }
        }
        points.push(code);
    }
    reader.x = x;
    return fromCharCode.apply(String, points);
};
var JsonDecoder = (function () {
    function JsonDecoder() {
        this.reader = new Reader_1.Reader();
    }
    JsonDecoder.prototype.read = function (uint8) {
        this.reader.reset(uint8);
        return this.readAny();
    };
    JsonDecoder.prototype.readAny = function () {
        this.skipWhitespace();
        var reader = this.reader;
        var x = reader.x;
        var uint8 = reader.uint8;
        var char = uint8[x];
        switch (char) {
            case 34:
                return uint8[x + 1] === 0x64
                    ? this.tryReadBin() || this.readStr()
                    : this.readStr();
            case 91:
                return this.readArr();
            case 102:
                return this.readFalse();
            case 110:
                return this.readNull();
            case 116:
                return this.readTrue();
            case 123:
                return this.readObj();
            default:
                if ((char >= 48 && char <= 57) || char === 45)
                    return this.readNum();
                throw new Error('Invalid JSON');
        }
    };
    JsonDecoder.prototype.skipWhitespace = function () {
        var reader = this.reader;
        var uint8 = reader.uint8;
        var x = reader.x;
        var char = 0;
        while (true) {
            char = uint8[x];
            switch (char) {
                case 32:
                case 9:
                case 10:
                case 13:
                    x++;
                    continue;
                default:
                    reader.x = x;
                    return;
            }
        }
    };
    JsonDecoder.prototype.readNull = function () {
        if (this.reader.u32() !== 0x6e756c6c)
            throw new Error('Invalid JSON');
        return null;
    };
    JsonDecoder.prototype.readTrue = function () {
        if (this.reader.u32() !== 0x74727565)
            throw new Error('Invalid JSON');
        return true;
    };
    JsonDecoder.prototype.readFalse = function () {
        var reader = this.reader;
        if (reader.u8() !== 0x66 || reader.u32() !== 0x616c7365)
            throw new Error('Invalid JSON');
        return false;
    };
    JsonDecoder.prototype.readBool = function () {
        var reader = this.reader;
        switch (reader.uint8[reader.x]) {
            case 102:
                return this.readFalse();
            case 116:
                return this.readTrue();
            default:
                throw new Error('Invalid JSON');
        }
    };
    JsonDecoder.prototype.readNum = function () {
        var reader = this.reader;
        var uint8 = reader.uint8;
        var x = reader.x;
        var c = uint8[x++];
        var c1 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 43 && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c2 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 43 && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c3 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 43 && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c4 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 43 && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c5 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 43 && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c6 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 43 && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c7 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 43 && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c8 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 43 && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c9 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c10 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c11 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c12 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c13 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c14 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c15 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c16 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c17 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16, c17);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c18 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16, c17, c18);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c19 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16, c17, c18, c19);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c20 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16, c17, c18, c19, c20);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c21 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16, c17, c18, c19, c20, c21);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c22 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16, c17, c18, c19, c20, c21, c22);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c23 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16, c17, c18, c19, c20, c21, c22, c23);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        var c24 = c;
        c = uint8[x++];
        if (!c || ((c < 45 || c > 57) && c !== 69 && c !== 101)) {
            reader.x = x - 1;
            var num = +fromCharCode(c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16, c17, c18, c19, c20, c21, c22, c24);
            if (num !== num)
                throw new Error('Invalid JSON');
            return num;
        }
        throw new Error('Invalid JSON');
    };
    JsonDecoder.prototype.readStr = function () {
        var reader = this.reader;
        var uint8 = reader.uint8;
        var char = uint8[reader.x++];
        if (char !== 0x22)
            throw new Error('Invalid JSON');
        var x0 = reader.x;
        var x1 = findEndingQuote(uint8, x0);
        var str = (0, decodeUtf8_1.decodeUtf8)(uint8, x0, x1 - x0);
        str = str.replace(REGEX_REPLACE_ESCAPED_CHARS, escapedCharReplacer);
        reader.x = x1 + 1;
        return str;
    };
    JsonDecoder.prototype.tryReadBin = function () {
        var reader = this.reader;
        var u8 = reader.uint8;
        var x = reader.x;
        if (u8[x++] !== 0x22)
            return undefined;
        var hasDataUrlPrefix = hasBinaryPrefix(u8, x);
        if (!hasDataUrlPrefix)
            return undefined;
        x += 37;
        var x0 = x;
        x = findEndingQuote(u8, x);
        reader.x = x0;
        var bin = (0, fromBase64Bin_1.fromBase64Bin)(reader.view, x0, x - x0);
        reader.x = x + 1;
        return bin;
    };
    JsonDecoder.prototype.readBin = function () {
        var reader = this.reader;
        var u8 = reader.uint8;
        var x = reader.x;
        if (u8[x++] !== 0x22)
            throw new Error('Invalid JSON');
        var hasDataUrlPrefix = hasBinaryPrefix(u8, x);
        if (!hasDataUrlPrefix)
            throw new Error('Invalid JSON');
        x += 37;
        var x0 = x;
        x = findEndingQuote(u8, x);
        reader.x = x0;
        var bin = (0, fromBase64Bin_1.fromBase64Bin)(reader.view, x0, x - x0);
        reader.x = x + 1;
        return bin;
    };
    JsonDecoder.prototype.readArr = function () {
        var reader = this.reader;
        if (reader.u8() !== 0x5b)
            throw new Error('Invalid JSON');
        var arr = [];
        var uint8 = reader.uint8;
        while (true) {
            this.skipWhitespace();
            var char = uint8[reader.x];
            if (char === 0x5d)
                return reader.x++, arr;
            if (char === 0x2c) {
                reader.x++;
                continue;
            }
            arr.push(this.readAny());
        }
    };
    JsonDecoder.prototype.readObj = function () {
        var reader = this.reader;
        if (reader.u8() !== 0x7b)
            throw new Error('Invalid JSON');
        var obj = {};
        var uint8 = reader.uint8;
        while (true) {
            this.skipWhitespace();
            var char = uint8[reader.x];
            if (char === 0x7d)
                return reader.x++, obj;
            if (char === 0x2c) {
                reader.x++;
                continue;
            }
            char = uint8[reader.x++];
            if (char !== 0x22)
                throw new Error('Invalid JSON');
            var key = readShortUtf8StrAndUnescape(reader);
            if (key === '__proto__')
                throw new Error('Invalid JSON');
            this.skipWhitespace();
            if (reader.u8() !== 0x3a)
                throw new Error('Invalid JSON');
            this.skipWhitespace();
            obj[key] = this.readAny();
        }
    };
    return JsonDecoder;
}());
exports.JsonDecoder = JsonDecoder;
