// json-joy/lib/reactive-rpc/common/codec/compact/CompactRpcMessageCodec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompactRpcMessageCodec = void 0;
var tslib_1 = require("tslib");
var error_1 = require("../../rpc/caller/error");
var msg = tslib_1.__importStar(require("../../messages"));
var Value_1 = require("../../messages/Value");
var CborEncoder_1 = require("../../../../json-pack/cbor/CborEncoder");
var msgpack_1 = require("../../../../json-pack/msgpack");
var JsonEncoder_1 = require("../../../../json-pack/json/JsonEncoder");
var fromJson = function (arr) {
    if (!(arr instanceof Array))
        throw error_1.RpcError.fromCode(error_1.RpcErrorCodes.BAD_REQUEST);
    var type = arr[0];
    switch (type) {
        case 1: {
            var data = arr[3];
            var value = data === undefined ? data : new Value_1.Value(data, undefined);
            return new msg.RequestCompleteMessage(arr[1], arr[2], value);
        }
        case 0: {
            var data = arr[3];
            var value = data === undefined ? data : new Value_1.Value(data, undefined);
            return new msg.RequestDataMessage(arr[1], arr[2], value);
        }
        case 2: {
            return new msg.RequestErrorMessage(arr[1], arr[2], new Value_1.Value(arr[3], undefined));
        }
        case 3: {
            return new msg.RequestUnsubscribeMessage(arr[1]);
        }
        case 5: {
            var data = arr[2];
            var value = data === undefined ? data : new Value_1.Value(data, undefined);
            return new msg.ResponseCompleteMessage(arr[1], value);
        }
        case 4: {
            return new msg.ResponseDataMessage(arr[1], new Value_1.Value(arr[2], undefined));
        }
        case 6: {
            return new msg.ResponseErrorMessage(arr[1], new Value_1.Value(arr[2], undefined));
        }
        case 7: {
            return new msg.ResponseUnsubscribeMessage(arr[1]);
        }
        case 8: {
            return new msg.NotificationMessage(arr[1], new Value_1.Value(arr[2], undefined));
        }
    }
    throw error_1.RpcError.value(error_1.RpcError.validation('Unknown message type'));
};
var CompactRpcMessageCodec = (function () {
    function CompactRpcMessageCodec() {
        this.id = 'rx.compact';
        this.format = 1;
    }
    CompactRpcMessageCodec.prototype.encodeMessage = function (jsonCodec, message) {
        message.encodeCompact(jsonCodec);
    };
    CompactRpcMessageCodec.prototype.encodeBatch = function (jsonCodec, batch) {
        var encoder = jsonCodec.encoder;
        if (encoder instanceof CborEncoder_1.CborEncoder || encoder instanceof msgpack_1.MsgPackEncoder) {
            var length_1 = batch.length;
            encoder.writeArrHdr(length_1);
            for (var i = 0; i < length_1; i++)
                batch[i].encodeCompact(jsonCodec);
        }
        else if (encoder instanceof JsonEncoder_1.JsonEncoder) {
            var length_2 = batch.length;
            var last = length_2 - 1;
            encoder.writeStartArr();
            for (var i = 0; i < last; i++) {
                batch[i].encodeCompact(jsonCodec);
                encoder.writeArrSeparator();
            }
            if (length_2 > 0)
                batch[last].encodeCompact(jsonCodec);
            encoder.writeEndArr();
        }
        else {
            var jsonMessages = [];
            var length_3 = batch.length;
            for (var i = 0; i < length_3; i++)
                jsonMessages.push(batch[i].toCompact());
            var encoder_1 = jsonCodec.encoder;
            encoder_1.writeArr(jsonMessages);
        }
    };
    CompactRpcMessageCodec.prototype.encode = function (jsonCodec, batch) {
        var encoder = jsonCodec.encoder;
        var writer = encoder.writer;
        writer.reset();
        this.encodeBatch(jsonCodec, batch);
        return writer.flush();
    };
    CompactRpcMessageCodec.prototype.decodeBatch = function (jsonCodec, uint8) {
        var decoder = jsonCodec.decoder;
        var value = decoder.read(uint8);
        if (!(value instanceof Array))
            throw error_1.RpcError.invalidRequest();
        if (typeof value[0] === 'number')
            return [fromJson(value)];
        var result = [];
        var length = value.length;
        for (var i = 0; i < length; i++) {
            var item = value[i];
            result.push(fromJson(item));
        }
        return result;
    };
    CompactRpcMessageCodec.prototype.fromJson = function (compact) {
        return fromJson(compact);
    };
    return CompactRpcMessageCodec;
}());
exports.CompactRpcMessageCodec = CompactRpcMessageCodec;
