// json-joy/es6/json-crdt/types/rga-string/StringRga.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StringRga = exports.StringChunk = void 0;
const clock_1 = require("../../../json-crdt-patch/clock");
const AbstractRga_1 = require("../rga/AbstractRga");
class StringChunk {
    constructor(id, span, str) {
        this.id = id;
        this.span = span;
        this.len = str ? span : 0;
        this.del = !str;
        this.p = undefined;
        this.l = undefined;
        this.r = undefined;
        this.p2 = undefined;
        this.l2 = undefined;
        this.r2 = undefined;
        this.s = undefined;
        this.data = str;
    }
    merge(str) {
        this.data += str;
        this.span = this.data.length;
    }
    split(ticks) {
        if (!this.del) {
            const chunk = new StringChunk((0, clock_1.tick)(this.id, ticks), this.span - ticks, this.data.slice(ticks));
            this.data = this.data.slice(0, ticks);
            this.span = ticks;
            return chunk;
        }
        const chunk = new StringChunk((0, clock_1.tick)(this.id, ticks), this.span - ticks, '');
        this.span = ticks;
        return chunk;
    }
    delete() {
        this.del = true;
        this.data = '';
    }
    clone() {
        const chunk = new StringChunk(this.id, this.span, this.data);
        return chunk;
    }
}
exports.StringChunk = StringChunk;
class StringRga extends AbstractRga_1.AbstractRga {
    constructor() {
        super(...arguments);
        this._view = '';
    }
    create() {
        return new StringRga(this.id);
    }
    createChunk(id, str) {
        return new StringChunk(id, str ? str.length : 0, str || '');
    }
    onViewChange() {
        this._view = '';
    }
    view() {
        if (this._view)
            return this._view;
        let str = '';
        for (let chunk = this.first(); chunk; chunk = this.next(chunk))
            str += chunk.data;
        return (this._view = str);
    }
    children() { }
    child() {
        return undefined;
    }
    container() {
        return undefined;
    }
}
exports.StringRga = StringRga;
