// json-joy/lib/json-pack/ion/ast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toAst = exports.AnnotationAstNode = exports.ObjAstNode = exports.ArrAstNode = exports.BinAstNode = exports.StrAstNode = exports.FloatAstNode = exports.NintAstNode = exports.UintAstNode = exports.BoolAstNode = exports.NullAstNode = void 0;
var utf8_1 = require("../../util/strings/utf8");
var NullAstNode = (function () {
    function NullAstNode() {
        this.val = null;
        this.len = 1;
    }
    NullAstNode.prototype.byteLength = function () {
        return 1;
    };
    return NullAstNode;
}());
exports.NullAstNode = NullAstNode;
var BoolAstNode = (function () {
    function BoolAstNode(val) {
        this.val = val;
        this.len = 1;
    }
    BoolAstNode.prototype.byteLength = function () {
        return 1;
    };
    return BoolAstNode;
}());
exports.BoolAstNode = BoolAstNode;
var UintAstNode = (function () {
    function UintAstNode(val) {
        this.val = val;
        if (!val)
            this.len = 0;
        else if (val <= 0xff)
            this.len = 1;
        else if (val <= 0xffff)
            this.len = 2;
        else if (val <= 0xffffff)
            this.len = 3;
        else if (val <= 0xffffffff)
            this.len = 4;
        else if (val <= 0xffffffffff)
            this.len = 5;
        else if (val <= 0xffffffffffff)
            this.len = 6;
        else
            this.len = 7;
    }
    UintAstNode.prototype.byteLength = function () {
        return 1 + this.len;
    };
    return UintAstNode;
}());
exports.UintAstNode = UintAstNode;
var NintAstNode = (function () {
    function NintAstNode(val) {
        this.val = val;
        var uint = -val;
        if (!uint)
            this.len = 0;
        else if (uint <= 0xff)
            this.len = 1;
        else if (uint <= 0xffff)
            this.len = 2;
        else if (uint <= 0xffffff)
            this.len = 3;
        else if (uint <= 0xffffffff)
            this.len = 4;
        else if (uint <= 0xffffffffff)
            this.len = 5;
        else if (uint <= 0xffffffffffff)
            this.len = 6;
        else
            this.len = 7;
    }
    NintAstNode.prototype.byteLength = function () {
        return 1 + this.len;
    };
    return NintAstNode;
}());
exports.NintAstNode = NintAstNode;
var FloatAstNode = (function () {
    function FloatAstNode(val) {
        this.val = val;
        this.len = 8;
    }
    FloatAstNode.prototype.byteLength = function () {
        return 1 + this.len;
    };
    return FloatAstNode;
}());
exports.FloatAstNode = FloatAstNode;
var vUintLen = function (num) {
    if (num <= 127)
        return 1;
    else if (num <= 16383)
        return 2;
    else if (num <= 2097151)
        return 3;
    else if (num <= 268435455)
        return 4;
    else if (num <= 34359738367)
        return 5;
    else
        return 6;
};
var StrAstNode = (function () {
    function StrAstNode(val) {
        this.val = val;
        this.len = (0, utf8_1.utf8Count)(val);
    }
    StrAstNode.prototype.byteLength = function () {
        return this.len < 14 ? 1 + this.len : 1 + vUintLen(this.len) + this.len;
    };
    return StrAstNode;
}());
exports.StrAstNode = StrAstNode;
var BinAstNode = (function () {
    function BinAstNode(val) {
        this.val = val;
        this.len = val.length;
    }
    BinAstNode.prototype.byteLength = function () {
        return this.len < 14 ? 1 + this.len : 1 + vUintLen(this.len) + this.len;
    };
    return BinAstNode;
}());
exports.BinAstNode = BinAstNode;
var ArrAstNode = (function () {
    function ArrAstNode(val) {
        this.val = val;
        if (val === null) {
            this.len = 1;
        }
        else {
            if (!val.length)
                this.len = 0;
            else {
                var elementLength = 0;
                for (var i = 0; i < val.length; i++)
                    elementLength += val[i].byteLength();
                this.len = elementLength;
            }
        }
    }
    ArrAstNode.prototype.byteLength = function () {
        return this.len < 14 ? 1 + this.len : 1 + vUintLen(this.len) + this.len;
    };
    return ArrAstNode;
}());
exports.ArrAstNode = ArrAstNode;
var ObjAstNode = (function () {
    function ObjAstNode(val) {
        this.val = val;
        if (val === null) {
            this.len = 1;
        }
        else {
            if (!val.size)
                this.len = 0;
            else {
                var len_1 = 0;
                val.forEach(function (node, symbolId) {
                    len_1 += vUintLen(symbolId) + node.byteLength();
                });
                this.len = len_1;
            }
        }
    }
    ObjAstNode.prototype.byteLength = function () {
        return this.len < 14 ? 1 + this.len : 1 + vUintLen(this.len) + this.len;
    };
    return ObjAstNode;
}());
exports.ObjAstNode = ObjAstNode;
var AnnotationAstNode = (function () {
    function AnnotationAstNode(val, annotations) {
        this.val = val;
        this.annotations = annotations;
        var len = 0;
        for (var i = 0; i < annotations.length; i++)
            len += vUintLen(annotations[i]);
        this.annotationLen = len;
        len += vUintLen(len);
        len += val.byteLength();
        this.len = len;
    }
    AnnotationAstNode.prototype.byteLength = function () {
        return this.len < 14 ? 1 + this.len : 1 + vUintLen(this.len) + this.len;
    };
    return AnnotationAstNode;
}());
exports.AnnotationAstNode = AnnotationAstNode;
var isSafeInteger = Number.isSafeInteger;
var toAst = function (val, symbols) {
    if (val === null)
        return new NullAstNode();
    if (val instanceof Array)
        return new ArrAstNode(val.map(function (el) { return (0, exports.toAst)(el, symbols); }));
    if (val instanceof Uint8Array)
        return new BinAstNode(val);
    switch (typeof val) {
        case 'boolean':
            return new BoolAstNode(val);
        case 'number': {
            if (isSafeInteger(val))
                return val >= 0 ? new UintAstNode(val) : new NintAstNode(val);
            else
                return new FloatAstNode(val);
        }
        case 'string':
            return new StrAstNode(val);
        case 'object': {
            var struct = new Map();
            for (var key in val) {
                var symbolId = symbols.add(key);
                struct.set(symbolId, (0, exports.toAst)(val[key], symbols));
            }
            return new ObjAstNode(struct);
        }
    }
    throw new Error('UNKNOWN_TYPE');
};
exports.toAst = toAst;
