// json-joy/lib/json-pack/ion/Import.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Import = void 0;
var ast_1 = require("./ast");
var Import = (function () {
    function Import(parent, symbols) {
        this.parent = parent;
        this.symbols = symbols;
        this.byText = new Map();
        this.offset = parent ? parent.offset + parent.length : 1;
        this.length = symbols.length;
        for (var i = 0; i < symbols.length; i++) {
            var symbol = symbols[i];
            this.byText.set(symbol, this.offset + i);
        }
    }
    Import.prototype.getId = function (symbol) {
        var id = this.byText.get(symbol);
        if (id !== undefined)
            return id;
        if (this.parent)
            this.parent.getId(symbol);
        return undefined;
    };
    Import.prototype.getText = function (id) {
        if (id < this.offset)
            return this.parent ? this.parent.getText(id) : undefined;
        return this.symbols[id - this.offset];
    };
    Import.prototype.add = function (symbol) {
        var id = this.byText.get(symbol);
        if (id !== undefined)
            return id;
        var length = this.symbols.length;
        id = this.offset + length;
        this.symbols.push(symbol);
        this.length++;
        this.byText.set(symbol, id);
        return id;
    };
    Import.prototype.toAst = function () {
        var map = new Map();
        map.set(7, (0, ast_1.toAst)(this.symbols, this));
        return new ast_1.ObjAstNode(map);
    };
    return Import;
}());
exports.Import = Import;
