// json-joy/lib/json-pack/ion/IonEncoderFast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IonEncoderFast = void 0;
var Writer_1 = require("../../util/buffers/Writer");
var ast_1 = require("./ast");
var Import_1 = require("./Import");
var symbols_1 = require("./symbols");
var IonEncoderFast = (function () {
    function IonEncoderFast(writer) {
        if (writer === void 0) { writer = new Writer_1.Writer(); }
        this.writer = writer;
    }
    IonEncoderFast.prototype.encode = function (value) {
        this.writer.reset();
        this.symbols = new Import_1.Import(symbols_1.systemSymbolImport, []);
        var ast = (0, ast_1.toAst)(value, this.symbols);
        this.writeIvm();
        this.writeSymbolTable();
        this.writeAny(ast);
        return this.writer.flush();
    };
    IonEncoderFast.prototype.writeAny = function (value) {
        if (value instanceof ast_1.NullAstNode)
            this.writer.u8(0 + 15);
        else if (value instanceof ast_1.StrAstNode)
            this.writeStr(value);
        else if (value instanceof ast_1.UintAstNode)
            this.encodeUint(value);
        else if (value instanceof ast_1.NintAstNode)
            this.encodeNint(value);
        else if (value instanceof ast_1.ObjAstNode)
            this.writeObj(value);
        else if (value instanceof ast_1.ArrAstNode)
            this.writeArr(value);
        else if (value instanceof ast_1.FloatAstNode)
            this.writeFloat(value);
        else if (value instanceof ast_1.BoolAstNode)
            this.writeBool(value);
        else if (value instanceof ast_1.BinAstNode)
            this.writeBin(value);
    };
    IonEncoderFast.prototype.writeIvm = function () {
        this.writer.u32(0xe00100ea);
    };
    IonEncoderFast.prototype.writeSymbolTable = function () {
        var _a;
        if (!((_a = this.symbols) === null || _a === void 0 ? void 0 : _a.length))
            return;
        var node = new ast_1.AnnotationAstNode(this.symbols.toAst(), [3]);
        this.writeAnnotations(node);
    };
    IonEncoderFast.prototype.writeAnnotations = function (node) {
        var writer = this.writer;
        if (node.len < 14)
            writer.u8(224 + node.len);
        else {
            writer.u8(224 + 14);
            this.writeVUint(node.len);
        }
        this.writeVUint(node.annotationLen);
        for (var i = 0; i < node.annotations.length; i++)
            this.writeVUint(node.annotations[i]);
        this.writeAny(node.val);
    };
    IonEncoderFast.prototype.writeBool = function (node) {
        this.writer.u8(16 + (node.val ? 1 : 0));
    };
    IonEncoderFast.prototype.encodeUint = function (node) {
        var uint = node.val;
        if (!uint)
            this.writer.u8(32);
        else if (uint <= 0xff)
            this.writer.u16(((32 + 1) << 8) + uint);
        else if (uint <= 0xffff)
            this.writer.u8u16(32 + 2, uint);
        else if (uint <= 0xffffff)
            this.writer.u32(((32 + 3) << 24) + uint);
        else if (uint <= 0xffffffff)
            this.writer.u8u32(32 + 4, uint);
        else {
            var lo = uint | 0;
            if (lo < 0)
                lo += 4294967296;
            var hi = uint - lo;
            hi /= 4294967296;
            if (uint <= 0xffffffffff) {
                this.writer.u16(((32 + 5) << 8) + hi);
                this.writer.u32(lo);
            }
            else if (uint <= 0xffffffffffff) {
                this.writer.u8u16(32 + 6, hi);
                this.writer.u32(lo);
            }
            else {
                this.writer.u16(((32 + 7) << 8) + (hi >> 16));
                this.writer.u16(hi & 0xffff);
                this.writer.u32(lo);
            }
        }
    };
    IonEncoderFast.prototype.encodeNint = function (node) {
        var uint = -node.val;
        if (uint <= 0xff)
            this.writer.u16(((48 + 1) << 8) + uint);
        else if (uint <= 0xffff)
            this.writer.u8u16(48 + 2, uint);
        else if (uint <= 0xffffff)
            this.writer.u32(((48 + 3) << 24) + uint);
        else if (uint <= 0xffffffff)
            this.writer.u8u32(48 + 4, uint);
        else {
            var lo = uint | 0;
            if (lo < 0)
                lo += 4294967296;
            var hi = uint - lo;
            hi /= 4294967296;
            if (uint <= 0xffffffffff) {
                this.writer.u16(((48 + 5) << 8) + hi);
                this.writer.u32(lo);
            }
            else if (uint <= 0xffffffffffff) {
                this.writer.u8u16(48 + 6, hi);
                this.writer.u32(lo);
            }
            else {
                this.writer.u16(((48 + 7) << 8) + (hi >> 16));
                this.writer.u16(hi & 0xffff);
                this.writer.u32(lo);
            }
        }
    };
    IonEncoderFast.prototype.writeFloat = function (node) {
        this.writer.u8f64(64 + 8, node.val);
    };
    IonEncoderFast.prototype.writeVUint = function (num) {
        var writer = this.writer;
        if (num <= 127) {
            writer.u8(128 + num);
        }
        else if (num <= 16383) {
            writer.ensureCapacity(2);
            var uint8 = writer.uint8;
            uint8[writer.x++] = num >>> 7;
            uint8[writer.x++] = 128 + (num & 127);
        }
        else if (num <= 2097151) {
            writer.ensureCapacity(3);
            var uint8 = writer.uint8;
            uint8[writer.x++] = num >>> 14;
            uint8[writer.x++] = (num >>> 7) & 127;
            uint8[writer.x++] = 128 + (num & 127);
        }
        else if (num <= 268435455) {
            writer.ensureCapacity(4);
            var uint8 = writer.uint8;
            uint8[writer.x++] = num >>> 21;
            uint8[writer.x++] = (num >>> 14) & 127;
            uint8[writer.x++] = (num >>> 7) & 127;
            uint8[writer.x++] = 128 + (num & 127);
        }
        else {
            var lo32 = num | 0;
            if (lo32 < 0)
                lo32 += 4294967296;
            var hi32 = (num - lo32) / 4294967296;
            if (num <= 34359738367) {
                writer.ensureCapacity(5);
                var uint8 = writer.uint8;
                uint8[writer.x++] = (hi32 << 4) | (num >>> 28);
                uint8[writer.x++] = (num >>> 21) & 127;
                uint8[writer.x++] = (num >>> 14) & 127;
                uint8[writer.x++] = (num >>> 7) & 127;
                uint8[writer.x++] = 128 + (num & 127);
            }
            else if (num <= 4398046511103) {
                writer.ensureCapacity(6);
                var uint8 = writer.uint8;
                uint8[writer.x++] = (hi32 >>> 3) & 15;
                uint8[writer.x++] = ((hi32 & 7) << 4) | (num >>> 28);
                uint8[writer.x++] = (num >>> 21) & 127;
                uint8[writer.x++] = (num >>> 14) & 127;
                uint8[writer.x++] = (num >>> 7) & 127;
                uint8[writer.x++] = 128 + (num & 127);
            }
        }
    };
    IonEncoderFast.prototype.writeStr = function (node) {
        var str = node.val;
        var length = node.len;
        var writer = this.writer;
        if (length < 14)
            writer.u8(128 + length);
        else {
            writer.u8(128 + 14);
            this.writeVUint(length);
        }
        writer.utf8(str);
    };
    IonEncoderFast.prototype.writeBin = function (node) {
        var buf = node.val;
        var length = node.len;
        var writer = this.writer;
        if (length < 14)
            writer.u8(160 + length);
        else {
            writer.u8(160 + 14);
            this.writeVUint(length);
        }
        writer.buf(buf, length);
    };
    IonEncoderFast.prototype.writeArr = function (node) {
        var writer = this.writer;
        var arr = node.val;
        if (arr === null) {
            writer.u8(176 + 15);
            return;
        }
        var length = node.len;
        if (length < 14)
            writer.u8(176 + length);
        else {
            writer.u8(176 + 14);
            this.writeVUint(length);
        }
        for (var i = 0; i < length; i++)
            this.writeAny(arr[i]);
    };
    IonEncoderFast.prototype.writeObj = function (node) {
        var _this = this;
        var writer = this.writer;
        var arr = node.val;
        if (arr === null) {
            writer.u8(176 + 15);
            return;
        }
        var length = node.len;
        if (length < 14)
            writer.u8(208 + length);
        else {
            writer.u8(208 + 14);
            this.writeVUint(length);
        }
        node.val.forEach(function (n, symbolId) {
            _this.writeVUint(symbolId);
            _this.writeAny(n);
        });
    };
    return IonEncoderFast;
}());
exports.IonEncoderFast = IonEncoderFast;
