// json-joy/es2020/json-crdt-patch/PatchBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PatchBuilder = void 0;
const ArrInsOp_1 = require("./operations/ArrInsOp");
const ArrOp_1 = require("./operations/ArrOp");
const BinInsOp_1 = require("./operations/BinInsOp");
const BinOp_1 = require("./operations/BinOp");
const ConstOp_1 = require("./operations/ConstOp");
const DelOp_1 = require("./operations/DelOp");
const clock_1 = require("./clock");
const isUint8Array_1 = require("../util/buffers/isUint8Array");
const NoopOp_1 = require("./operations/NoopOp");
const ObjOp_1 = require("./operations/ObjOp");
const ObjSetOp_1 = require("./operations/ObjSetOp");
const Patch_1 = require("./Patch");
const StrInsOp_1 = require("./operations/StrInsOp");
const StrOp_1 = require("./operations/StrOp");
const ValOp_1 = require("./operations/ValOp");
const ValSetOp_1 = require("./operations/ValSetOp");
const TupOp_1 = require("./operations/TupOp");
const constants_1 = require("./constants");
const Tuple_1 = require("./builder/Tuple");
const Konst_1 = require("./builder/Konst");
const DelayedValueBuilder_1 = require("./builder/DelayedValueBuilder");
const maybeConst = (x) => {
    switch (typeof x) {
        case 'number':
        case 'boolean':
            return true;
        default:
            return x === null;
    }
};
class PatchBuilder {
    constructor(clock) {
        this.clock = clock;
        this.patch = new Patch_1.Patch();
    }
    nextTime() {
        return this.patch.nextTime() || this.clock.time;
    }
    flush() {
        const patch = this.patch;
        this.patch = new Patch_1.Patch();
        return patch;
    }
    obj() {
        this.pad();
        const id = this.clock.tick(1);
        this.patch.ops.push(new ObjOp_1.ObjOp(id));
        return id;
    }
    arr() {
        this.pad();
        const id = this.clock.tick(1);
        this.patch.ops.push(new ArrOp_1.ArrOp(id));
        return id;
    }
    tup() {
        this.pad();
        const id = this.clock.tick(1);
        this.patch.ops.push(new TupOp_1.TupOp(id));
        return id;
    }
    str() {
        this.pad();
        const id = this.clock.tick(1);
        this.patch.ops.push(new StrOp_1.StrOp(id));
        return id;
    }
    bin() {
        this.pad();
        const id = this.clock.tick(1);
        this.patch.ops.push(new BinOp_1.BinOp(id));
        return id;
    }
    const(value) {
        this.pad();
        const id = this.clock.tick(1);
        this.patch.ops.push(new ConstOp_1.ConstOp(id, value));
        return id;
    }
    val(val) {
        this.pad();
        const id = this.clock.tick(1);
        this.patch.ops.push(new ValOp_1.ValOp(id, val));
        return id;
    }
    root(val) {
        this.pad();
        const id = this.clock.tick(1);
        this.patch.ops.push(new ValSetOp_1.ValSetOp(id, constants_1.ORIGIN, val));
        return id;
    }
    setKeys(obj, data) {
        this.pad();
        if (!data.length)
            throw new Error('EMPTY_TUPLES');
        const id = this.clock.tick(0);
        const op = new ObjSetOp_1.ObjSetOp(id, obj, data);
        const span = op.span();
        if (span > 1)
            this.clock.tick(span - 1);
        this.patch.ops.push(op);
        return id;
    }
    setVal(obj, val) {
        this.pad();
        const id = this.clock.tick(0);
        const op = new ValSetOp_1.ValSetOp(id, obj, val);
        this.patch.ops.push(op);
        return id;
    }
    insStr(obj, ref, data) {
        this.pad();
        if (!data.length)
            throw new Error('EMPTY_STRING');
        const id = this.clock.tick(1);
        const op = new StrInsOp_1.StrInsOp(id, obj, ref, data);
        const span = op.span();
        if (span > 1)
            this.clock.tick(span - 1);
        this.patch.ops.push(op);
        return id;
    }
    insBin(obj, ref, data) {
        this.pad();
        if (!data.length)
            throw new Error('EMPTY_BINARY');
        const id = this.clock.tick(1);
        const op = new BinInsOp_1.BinInsOp(id, obj, ref, data);
        const span = op.span();
        if (span > 1)
            this.clock.tick(span - 1);
        this.patch.ops.push(op);
        return id;
    }
    insArr(arr, ref, data) {
        this.pad();
        const id = this.clock.tick(1);
        const op = new ArrInsOp_1.ArrInsOp(id, arr, ref, data);
        const span = op.span();
        if (span > 1)
            this.clock.tick(span - 1);
        this.patch.ops.push(op);
        return id;
    }
    del(obj, what) {
        this.pad();
        const id = this.clock.tick(0);
        this.patch.ops.push(new DelOp_1.DelOp(id, obj, what));
        return id;
    }
    noop(span) {
        this.pad();
        const id = this.clock.tick(span);
        this.patch.ops.push(new NoopOp_1.NoopOp(id, span));
        return id;
    }
    jsonObj(json) {
        const obj = this.obj();
        const keys = Object.keys(json);
        if (keys.length) {
            const tuples = [];
            for (const k of keys) {
                const value = json[k];
                const valueId = value instanceof clock_1.Timestamp ? value : maybeConst(value) ? this.const(value) : this.json(value);
                tuples.push([k, valueId]);
            }
            this.setKeys(obj, tuples);
        }
        return obj;
    }
    jsonArr(json) {
        const arr = this.arr();
        if (json.length) {
            const values = [];
            for (const el of json)
                values.push(this.json(el));
            this.insArr(arr, arr, values);
        }
        return arr;
    }
    jsonStr(json) {
        const str = this.str();
        if (json)
            this.insStr(str, str, json);
        return str;
    }
    jsonBin(json) {
        const bin = this.bin();
        if (json.length)
            this.insBin(bin, bin, json);
        return bin;
    }
    jsonVal(json) {
        const val = this.const(json);
        return this.val(val);
    }
    jsonTup(slots) {
        const tup = this.tup();
        const length = slots.length;
        if (length) {
            const writes = [];
            for (let i = 0; i < length; i++)
                writes.push([i, this.constOrJson(slots[i])]);
            this.setKeys(tup, writes);
        }
        return tup;
    }
    json(json) {
        if (json instanceof clock_1.Timestamp)
            return json;
        if (json === undefined)
            return this.const(json);
        if (json instanceof Array)
            return this.jsonArr(json);
        if ((0, isUint8Array_1.isUint8Array)(json))
            return this.jsonBin(json);
        if (json instanceof Tuple_1.Tuple)
            return this.jsonTup(json.slots);
        if (json instanceof Konst_1.Konst)
            return this.const(json.val);
        if (json instanceof DelayedValueBuilder_1.DelayedValueBuilder)
            return json.build(this);
        switch (typeof json) {
            case 'object':
                return json === null ? this.jsonVal(json) : this.jsonObj(json);
            case 'string':
                return this.jsonStr(json);
            case 'number':
            case 'boolean':
                return this.jsonVal(json);
        }
        throw new Error('INVALID_JSON');
    }
    constOrJson(value) {
        if (value instanceof clock_1.Timestamp)
            return value;
        return maybeConst(value) ? this.const(value) : this.json(value);
    }
    maybeConst(value) {
        return value instanceof clock_1.Timestamp ? value : this.const(value);
    }
    pad() {
        const nextTime = this.patch.nextTime();
        if (!nextTime)
            return;
        const drift = this.clock.time - nextTime;
        if (drift > 0) {
            const id = (0, clock_1.ts)(this.clock.sid, nextTime);
            const padding = new NoopOp_1.NoopOp(id, drift);
            this.patch.ops.push(padding);
        }
    }
}
exports.PatchBuilder = PatchBuilder;
