// json-joy/lib/util/router/codegen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouterCodegenOpts = exports.RouterCodegenCtx = void 0;
var codegen_1 = require("../codegen");
var JsExpression_1 = require("../codegen/util/JsExpression");
var RouterCodegenCtx = (function () {
    function RouterCodegenCtx() {
        this.codegen = new codegen_1.Codegen({
            args: ['str'],
            prologue: 'str = "" + str; var len = str.length|0;',
            epilogue: 'return undefined;',
        });
    }
    return RouterCodegenCtx;
}());
exports.RouterCodegenCtx = RouterCodegenCtx;
var RouterCodegenOpts = (function () {
    function RouterCodegenOpts(slice, offset, depth) {
        if (depth === void 0) { depth = 0; }
        this.slice = slice;
        this.offset = offset;
        this.depth = depth;
    }
    RouterCodegenOpts.prototype.create = function (offset) {
        var slice = new JsExpression_1.JsExpression(function () { return "str.slice(".concat(offset, ")"); });
        return new RouterCodegenOpts(slice, offset, this.depth + 1);
    };
    return RouterCodegenOpts;
}());
exports.RouterCodegenOpts = RouterCodegenOpts;
