// json-joy/lib/util/router/steps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegexStep = exports.UntilStep = exports.ExactStep = void 0;
var ExactStep = (function () {
    function ExactStep(text) {
        this.text = text;
    }
    ExactStep.prototype.toText = function () {
        return this.text;
    };
    return ExactStep;
}());
exports.ExactStep = ExactStep;
var UntilStep = (function () {
    function UntilStep(name, until) {
        this.name = name;
        this.until = until;
    }
    UntilStep.prototype.toText = function () {
        var until = this.until === '\n' ? '\\n' : this.until;
        return "{".concat(this.name, "::").concat(until, "}");
    };
    return UntilStep;
}());
exports.UntilStep = UntilStep;
var RegexStep = (function () {
    function RegexStep(name, regex, until) {
        this.name = name;
        this.regex = regex;
        this.until = until;
    }
    RegexStep.prototype.toText = function () {
        var regex = this.regex || this.until ? ':' + this.regex : '';
        var until = this.until ? ':' + this.until : '';
        return "{".concat(this.name).concat(regex).concat(until, "}");
    };
    return RegexStep;
}());
exports.RegexStep = RegexStep;
