// json-joy/lib/json-crdt/types/lww-value/ValueLww.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValueLww = void 0;
var clock_1 = require("../../../json-crdt-patch/clock");
var printTree_1 = require("../../../util/print/printTree");
var ValueLww = (function () {
    function ValueLww(doc, id, val) {
        this.doc = doc;
        this.id = id;
        this.val = val;
    }
    ValueLww.prototype.set = function (val) {
        if ((0, clock_1.compare)(val, this.val) <= 0 && this.val.sid !== 0)
            return;
        var oldVal = this.val;
        this.val = val;
        return oldVal;
    };
    ValueLww.prototype.node = function () {
        return this.child();
    };
    ValueLww.prototype.view = function () {
        return this.node().view();
    };
    ValueLww.prototype.children = function (callback) {
        callback(this.node());
    };
    ValueLww.prototype.child = function () {
        return this.doc.index.get(this.val);
    };
    ValueLww.prototype.container = function () {
        var child = this.node();
        return child ? child.container() : undefined;
    };
    ValueLww.prototype.toString = function (tab) {
        var _this = this;
        if (tab === void 0) { tab = ''; }
        var node = this.node();
        var header = this.constructor.name + ' ' + (0, clock_1.toDisplayString)(this.id);
        return header + (0, printTree_1.printTree)(tab, [function (tab) { return (node ? node.toString(tab) : (0, clock_1.toDisplayString)(_this.val)); }]);
    };
    return ValueLww;
}());
exports.ValueLww = ValueLww;
