// json-joy/lib/reactive-rpc/common/rpc/methods/StaticRpcMethod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticRpcMethod = void 0;
var tslib_1 = require("tslib");
var rxjs_1 = require("rxjs");
var StaticRpcMethod = (function () {
    function StaticRpcMethod(options) {
        var _this = this;
        this.isStreaming = false;
        var call = options.call, validate = options.validate, onPreCall = options.onPreCall, pretty = options.pretty, req = options.req, res = options.res;
        this.pretty = !!pretty;
        this.validate = validate;
        this.onPreCall = onPreCall;
        this.req = req;
        this.res = res;
        var responseIsVoid = !!res && res.getTypeName() === 'const' && res.getSchema().value === undefined;
        this.acceptsNotifications = responseIsVoid;
        this.call = call;
        this.call$ = function (request$, ctx) { return (0, rxjs_1.from)((function () { return tslib_1.__awaiter(_this, void 0, void 0, function () { var _a; return tslib_1.__generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = this.call;
                    return [4, (0, rxjs_1.firstValueFrom)(request$)];
                case 1: return [2, _a.apply(this, [_b.sent(), ctx])];
            }
        }); }); })()); };
    }
    return StaticRpcMethod;
}());
exports.StaticRpcMethod = StaticRpcMethod;
