// json-joy/lib/json-crdt-patch/util/binary/CrdtDecoder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrdtDecoder = void 0;
var tslib_1 = require("tslib");
var Reader_1 = require("../../../util/buffers/Reader");
var CrdtDecoder = (function (_super) {
    tslib_1.__extends(CrdtDecoder, _super);
    function CrdtDecoder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CrdtDecoder.prototype.id = function () {
        var byte = this.u8();
        if (byte <= 127)
            return [byte >>> 4, byte & 15];
        this.x--;
        return [this.b1vu28()[1], this.vu39()];
    };
    CrdtDecoder.prototype.idSkip = function () {
        var byte = this.u8();
        if (byte <= 127)
            return;
        this.x--;
        this.b1vu28Skip();
        this.vu39Skip();
    };
    CrdtDecoder.prototype.vu57 = function () {
        var o1 = this.u8();
        if (o1 <= 127)
            return o1;
        var o2 = this.u8();
        if (o2 <= 127)
            return (o2 << 7) | (o1 & 127);
        var o3 = this.u8();
        if (o3 <= 127)
            return (o3 << 14) | ((o2 & 127) << 7) | (o1 & 127);
        var o4 = this.u8();
        if (o4 <= 127)
            return (o4 << 21) | ((o3 & 127) << 14) | ((o2 & 127) << 7) | (o1 & 127);
        var o5 = this.u8();
        if (o5 <= 127)
            return (o5 * 268435456 +
                (((o4 & 127) << 21) | ((o3 & 127) << 14) | ((o2 & 127) << 7) | (o1 & 127)));
        var o6 = this.u8();
        if (o6 <= 127)
            return (o6 * 34359738368 +
                ((o5 & 127) * 268435456 +
                    (((o4 & 127) << 21) | ((o3 & 127) << 14) | ((o2 & 127) << 7) | (o1 & 127))));
        var o7 = this.u8();
        if (o7 <= 127)
            return (o7 * 4398046511104 +
                ((o6 & 127) * 34359738368 +
                    ((o5 & 127) * 268435456 +
                        (((o4 & 127) << 21) | ((o3 & 127) << 14) | ((o2 & 127) << 7) | (o1 & 127)))));
        var o8 = this.u8();
        return (o8 * 562949953421312 +
            ((o7 & 127) * 4398046511104 +
                ((o6 & 127) * 34359738368 +
                    ((o5 & 127) * 268435456 +
                        (((o4 & 127) << 21) | ((o3 & 127) << 14) | ((o2 & 127) << 7) | (o1 & 127))))));
    };
    CrdtDecoder.prototype.vu57Skip = function () {
        var o1 = this.u8();
        if (o1 <= 127)
            return;
        var o2 = this.u8();
        if (o2 <= 127)
            return;
        var o3 = this.u8();
        if (o3 <= 127)
            return;
        var o4 = this.u8();
        if (o4 <= 127)
            return;
        var o5 = this.u8();
        if (o5 <= 127)
            return;
        var o6 = this.u8();
        if (o6 <= 127)
            return;
        var o7 = this.u8();
        if (o7 <= 127)
            return;
        this.x++;
    };
    CrdtDecoder.prototype.vu39 = function () {
        var o1 = this.u8();
        if (o1 <= 127)
            return o1;
        var o2 = this.u8();
        if (o2 <= 127)
            return (o2 << 7) | (o1 & 127);
        var o3 = this.u8();
        if (o3 <= 127)
            return (o3 << 14) | ((o2 & 127) << 7) | (o1 & 127);
        var o4 = this.u8();
        if (o4 <= 127)
            return (o4 << 21) | ((o3 & 127) << 14) | ((o2 & 127) << 7) | (o1 & 127);
        var o5 = this.u8();
        if (o5 <= 127)
            return (o5 * 268435456 +
                (((o4 & 127) << 21) | ((o3 & 127) << 14) | ((o2 & 127) << 7) | (o1 & 127)));
        var o6 = this.u8();
        return ((o6 & 15) * 34359738368 +
            ((o5 & 127) * 268435456 +
                (((o4 & 127) << 21) | ((o3 & 127) << 14) | ((o2 & 127) << 7) | (o1 & 127))));
    };
    CrdtDecoder.prototype.vu39Skip = function () {
        var o1 = this.u8();
        if (o1 <= 127)
            return;
        var o2 = this.u8();
        if (o2 <= 127)
            return;
        var o3 = this.u8();
        if (o3 <= 127)
            return;
        var o4 = this.u8();
        if (o4 <= 127)
            return;
        var o5 = this.u8();
        if (o5 <= 127)
            return;
        this.x++;
    };
    CrdtDecoder.prototype.b1vu56 = function () {
        var byte = this.u8();
        var flag = !!(byte & 128);
        var o1 = 127 & byte;
        if (o1 <= 63)
            return [flag, o1];
        var o2 = this.u8();
        if (o2 <= 127)
            return [flag, (o2 << 6) | (o1 & 63)];
        var o3 = this.u8();
        if (o3 <= 127)
            return [flag, (o3 << 13) | ((o2 & 127) << 6) | (o1 & 63)];
        var o4 = this.u8();
        if (o4 <= 127)
            return [flag, (o4 << 20) | ((o3 & 127) << 13) | ((o2 & 127) << 6) | (o1 & 63)];
        var o5 = this.u8();
        if (o5 <= 127)
            return [
                flag,
                o5 * 134217728 +
                    (((o4 & 127) << 20) | ((o3 & 127) << 13) | ((o2 & 127) << 6) | (o1 & 63)),
            ];
        var o6 = this.u8();
        if (o6 <= 127)
            return [
                flag,
                o6 * 17179869184 +
                    ((o5 & 127) * 134217728 +
                        (((o4 & 127) << 20) | ((o3 & 127) << 13) | ((o2 & 127) << 6) | (o1 & 63))),
            ];
        var o7 = this.u8();
        if (o7 <= 127)
            return [
                flag,
                o7 * 2199023255552 +
                    ((o6 & 127) * 17179869184 +
                        ((o5 & 127) * 134217728 +
                            (((o4 & 127) << 20) |
                                ((o3 & 127) << 13) |
                                ((o2 & 127) << 6) |
                                (o1 & 63)))),
            ];
        var o8 = this.u8();
        return [
            flag,
            o8 * 281474976710656 +
                ((o7 & 127) * 2199023255552 +
                    ((o6 & 127) * 17179869184 +
                        ((o5 & 127) * 134217728 +
                            (((o4 & 127) << 20) |
                                ((o3 & 127) << 13) |
                                ((o2 & 127) << 6) |
                                (o1 & 63))))),
        ];
    };
    CrdtDecoder.prototype.b1vu28 = function () {
        var byte = this.u8();
        var flag = !!(byte & 128);
        var o1 = 127 & byte;
        if (o1 <= 63)
            return [flag, o1];
        var o2 = this.u8();
        if (o2 <= 127)
            return [flag, (o2 << 6) | (o1 & 63)];
        var o3 = this.u8();
        if (o3 <= 127)
            return [flag, (o3 << 13) | ((o2 & 127) << 6) | (o1 & 63)];
        var o4 = this.u8();
        return [flag, (o4 << 20) | ((o3 & 127) << 13) | ((o2 & 127) << 6) | (o1 & 63)];
    };
    CrdtDecoder.prototype.b1vu28Skip = function () {
        var byte = this.u8();
        var o1 = 127 & byte;
        if (o1 <= 63)
            return;
        var o2 = this.u8();
        if (o2 <= 127)
            return;
        var o3 = this.u8();
        if (o3 <= 127)
            return;
        this.x++;
    };
    CrdtDecoder.prototype.u53vu39 = function () {
        var x32 = this.u32();
        var x16 = this.u16();
        var y = this.u8();
        var z8 = this.u8();
        var x = (((y >>> 3) << 16) | x16) * 0x100000000 + x32;
        var z = ((y & 3) << 8) | z8;
        if (!(y & 4))
            return [x, z];
        var o1 = this.u8();
        if (o1 <= 127)
            return [x, (o1 << 10) | z];
        var o2 = this.u8();
        if (o2 <= 127)
            return [x, (o2 << 17) | ((o1 & 127) << 10) | z];
        var o3 = this.u8();
        if (o3 <= 127)
            return [x, (o3 << 24) | (((o2 & 127) << 17) | ((o1 & 127) << 10) | z)];
        var o4 = this.u8();
        return [
            x,
            o4 * 2147483648 +
                (((o3 & 127) << 24) | (((o2 & 127) << 17) | ((o1 & 127) << 10) | z)),
        ];
    };
    return CrdtDecoder;
}(Reader_1.Reader));
exports.CrdtDecoder = CrdtDecoder;
