// json-joy/lib/json-crdt-patch/util/binary/CrdtEncoder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrdtWriter = void 0;
var tslib_1 = require("tslib");
var Writer_1 = require("../../../util/buffers/Writer");
var CrdtWriter = (function (_super) {
    tslib_1.__extends(CrdtWriter, _super);
    function CrdtWriter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CrdtWriter.prototype.id = function (x, y) {
        if (x <= 7 && y <= 15) {
            this.u8((x << 4) | y);
        }
        else {
            this.b1vu28(true, x);
            this.vu39(y);
        }
    };
    CrdtWriter.prototype.vu57 = function (num) {
        if (num <= 127) {
            this.u8(num);
        }
        else if (num <= 16383) {
            this.ensureCapacity(2);
            var uint8 = this.uint8;
            uint8[this.x++] = 128 | (num & 127);
            uint8[this.x++] = num >>> 7;
        }
        else if (num <= 2097151) {
            this.ensureCapacity(3);
            var uint8 = this.uint8;
            uint8[this.x++] = 128 | (num & 127);
            uint8[this.x++] = 128 | ((num >>> 7) & 127);
            uint8[this.x++] = num >>> 14;
        }
        else if (num <= 268435455) {
            this.ensureCapacity(4);
            var uint8 = this.uint8;
            uint8[this.x++] = 128 | (num & 127);
            uint8[this.x++] = 128 | ((num >>> 7) & 127);
            uint8[this.x++] = 128 | ((num >>> 14) & 127);
            uint8[this.x++] = num >>> 21;
        }
        else {
            var lo32 = num | 0;
            if (lo32 < 0)
                lo32 += 4294967296;
            var hi32 = (num - lo32) / 4294967296;
            if (num <= 34359738367) {
                this.ensureCapacity(5);
                var uint8 = this.uint8;
                uint8[this.x++] = 128 | (num & 127);
                uint8[this.x++] = 128 | ((num >>> 7) & 127);
                uint8[this.x++] = 128 | ((num >>> 14) & 127);
                uint8[this.x++] = 128 | ((num >>> 21) & 127);
                uint8[this.x++] = (hi32 << 4) | (num >>> 28);
            }
            else if (num <= 4398046511103) {
                this.ensureCapacity(6);
                var uint8 = this.uint8;
                uint8[this.x++] = 128 | (num & 127);
                uint8[this.x++] = 128 | ((num >>> 7) & 127);
                uint8[this.x++] = 128 | ((num >>> 14) & 127);
                uint8[this.x++] = 128 | ((num >>> 21) & 127);
                uint8[this.x++] = 128 | ((hi32 & 7) << 4) | (num >>> 28);
                uint8[this.x++] = hi32 >>> 3;
            }
            else if (num <= 562949953421311) {
                this.ensureCapacity(7);
                var uint8 = this.uint8;
                uint8[this.x++] = 128 | (num & 127);
                uint8[this.x++] = 128 | ((num >>> 7) & 127);
                uint8[this.x++] = 128 | ((num >>> 14) & 127);
                uint8[this.x++] = 128 | ((num >>> 21) & 127);
                uint8[this.x++] = 128 | ((hi32 & 7) << 4) | (num >>> 28);
                uint8[this.x++] = 128 | ((hi32 & 1016) >>> 3);
                uint8[this.x++] = hi32 >>> 10;
            }
            else {
                this.ensureCapacity(8);
                var uint8 = this.uint8;
                uint8[this.x++] = 128 | (num & 127);
                uint8[this.x++] = 128 | ((num >>> 7) & 127);
                uint8[this.x++] = 128 | ((num >>> 14) & 127);
                uint8[this.x++] = 128 | ((num >>> 21) & 127);
                uint8[this.x++] = 128 | ((hi32 & 7) << 4) | (num >>> 28);
                uint8[this.x++] = 128 | ((hi32 & 1016) >>> 3);
                uint8[this.x++] = 128 | ((hi32 & 130048) >>> 10);
                uint8[this.x++] = hi32 >>> 17;
            }
        }
    };
    CrdtWriter.prototype.vu39 = function (num) {
        if (num <= 127) {
            this.u8(num);
        }
        else if (num <= 16383) {
            this.ensureCapacity(2);
            var uint8 = this.uint8;
            uint8[this.x++] = 128 | (num & 127);
            uint8[this.x++] = num >>> 7;
        }
        else if (num <= 2097151) {
            this.ensureCapacity(3);
            var uint8 = this.uint8;
            uint8[this.x++] = 128 | (num & 127);
            uint8[this.x++] = 128 | ((num >>> 7) & 127);
            uint8[this.x++] = num >>> 14;
        }
        else if (num <= 268435455) {
            this.ensureCapacity(4);
            var uint8 = this.uint8;
            uint8[this.x++] = 128 | (num & 127);
            uint8[this.x++] = 128 | ((num >>> 7) & 127);
            uint8[this.x++] = 128 | ((num >>> 14) & 127);
            uint8[this.x++] = num >>> 21;
        }
        else {
            var lo32 = num | 0;
            if (lo32 < 0)
                lo32 += 4294967296;
            var hi32 = (num - lo32) / 4294967296;
            if (num <= 34359738367) {
                this.ensureCapacity(5);
                var uint8 = this.uint8;
                uint8[this.x++] = 128 | (num & 127);
                uint8[this.x++] = 128 | ((num >>> 7) & 127);
                uint8[this.x++] = 128 | ((num >>> 14) & 127);
                uint8[this.x++] = 128 | ((num >>> 21) & 127);
                uint8[this.x++] = (hi32 << 4) | (num >>> 28);
            }
            else if (num <= 4398046511103) {
                this.ensureCapacity(6);
                var uint8 = this.uint8;
                uint8[this.x++] = 128 | (num & 127);
                uint8[this.x++] = 128 | ((num >>> 7) & 127);
                uint8[this.x++] = 128 | ((num >>> 14) & 127);
                uint8[this.x++] = 128 | ((num >>> 21) & 127);
                uint8[this.x++] = 128 | ((hi32 & 7) << 4) | (num >>> 28);
                uint8[this.x++] = (hi32 >>> 3) & 15;
            }
        }
    };
    CrdtWriter.prototype.b1vu56 = function (flag, num) {
        if (num <= 63) {
            this.u8((flag ? 128 : 0) | num);
        }
        else {
            var firstByteMask = flag ? 192 : 64;
            if (num <= 8191) {
                this.ensureCapacity(2);
                var uint8 = this.uint8;
                uint8[this.x++] = firstByteMask | (num & 63);
                uint8[this.x++] = num >>> 6;
            }
            else if (num <= 1048575) {
                this.ensureCapacity(3);
                var uint8 = this.uint8;
                uint8[this.x++] = firstByteMask | (num & 63);
                uint8[this.x++] = 128 | ((num >>> 6) & 127);
                uint8[this.x++] = num >>> 13;
            }
            else if (num <= 134217727) {
                this.ensureCapacity(4);
                var uint8 = this.uint8;
                uint8[this.x++] = firstByteMask | (num & 63);
                uint8[this.x++] = 128 | ((num >>> 6) & 127);
                uint8[this.x++] = 128 | ((num >>> 13) & 127);
                uint8[this.x++] = num >>> 20;
            }
            else {
                var lo32 = num | 0;
                if (lo32 < 0)
                    lo32 += 4294967296;
                var hi32 = (num - lo32) / 4294967296;
                if (num <= 17179869183) {
                    this.ensureCapacity(5);
                    var uint8 = this.uint8;
                    uint8[this.x++] = firstByteMask | (num & 63);
                    uint8[this.x++] = 128 | ((num >>> 6) & 127);
                    uint8[this.x++] = 128 | ((num >>> 13) & 127);
                    uint8[this.x++] = 128 | ((num >>> 20) & 127);
                    uint8[this.x++] = (hi32 << 5) | (num >>> 27);
                }
                else if (num <= 2199023255551) {
                    this.ensureCapacity(6);
                    var uint8 = this.uint8;
                    uint8[this.x++] = firstByteMask | (num & 63);
                    uint8[this.x++] = 128 | ((num >>> 6) & 127);
                    uint8[this.x++] = 128 | ((num >>> 13) & 127);
                    uint8[this.x++] = 128 | ((num >>> 20) & 127);
                    uint8[this.x++] = 128 | ((hi32 & 3) << 5) | (num >>> 27);
                    uint8[this.x++] = hi32 >>> 2;
                }
                else if (num <= 281474976710655) {
                    this.ensureCapacity(7);
                    var uint8 = this.uint8;
                    uint8[this.x++] = firstByteMask | (num & 63);
                    uint8[this.x++] = 128 | ((num >>> 6) & 127);
                    uint8[this.x++] = 128 | ((num >>> 13) & 127);
                    uint8[this.x++] = 128 | ((num >>> 20) & 127);
                    uint8[this.x++] = 128 | ((hi32 & 3) << 5) | (num >>> 27);
                    uint8[this.x++] = 128 | ((hi32 & 508) >>> 2);
                    uint8[this.x++] = hi32 >>> 9;
                }
                else {
                    this.ensureCapacity(8);
                    var uint8 = this.uint8;
                    uint8[this.x++] = firstByteMask | (num & 63);
                    uint8[this.x++] = 128 | ((num >>> 6) & 127);
                    uint8[this.x++] = 128 | ((num >>> 13) & 127);
                    uint8[this.x++] = 128 | ((num >>> 20) & 127);
                    uint8[this.x++] = 128 | ((hi32 & 3) << 5) | (num >>> 27);
                    uint8[this.x++] = 128 | ((hi32 & 508) >>> 2);
                    uint8[this.x++] = 128 | ((hi32 & 65024) >>> 9);
                    uint8[this.x++] = hi32 >>> 16;
                }
            }
        }
    };
    CrdtWriter.prototype.b1vu28 = function (flag, num) {
        if (num <= 63) {
            this.u8((flag ? 128 : 0) | num);
        }
        else {
            var firstByteMask = flag ? 192 : 64;
            if (num <= 8191) {
                this.ensureCapacity(2);
                var uint8 = this.uint8;
                uint8[this.x++] = firstByteMask | (num & 63);
                uint8[this.x++] = num >>> 6;
            }
            else if (num <= 1048575) {
                this.ensureCapacity(3);
                var uint8 = this.uint8;
                uint8[this.x++] = firstByteMask | (num & 63);
                uint8[this.x++] = 128 | ((num >>> 6) & 127);
                uint8[this.x++] = num >>> 13;
            }
            else {
                this.ensureCapacity(4);
                var uint8 = this.uint8;
                uint8[this.x++] = firstByteMask | (num & 63);
                uint8[this.x++] = 128 | ((num >>> 6) & 127);
                uint8[this.x++] = 128 | ((num >>> 13) & 127);
                uint8[this.x++] = num >>> 20;
            }
        }
    };
    CrdtWriter.prototype.u53vu39 = function (x, z) {
        var x1 = x | 0;
        if (x1 < 0)
            x1 += 4294967296;
        var x2 = (x - x1) / 4294967296;
        var fiveXBits = (x2 >>> 16) << 3;
        var twoZBits = (z >>> 8) & 3;
        var zFitsIn10Bits = z <= 1023;
        this.ensureCapacity(8);
        var uint8 = this.uint8;
        var view = this.view;
        view.setUint32(this.x, x1);
        this.x += 4;
        view.setUint16(this.x, x2 & 0xffff);
        this.x += 2;
        uint8[this.x++] = fiveXBits | (zFitsIn10Bits ? 0 : 4) | twoZBits;
        uint8[this.x++] = z & 0xff;
        if (zFitsIn10Bits)
            return;
        if (z <= 131071) {
            this.u8(z >>> 10);
        }
        else if (z <= 16777215) {
            this.ensureCapacity(2);
            var uint8_1 = this.uint8;
            uint8_1[this.x++] = 128 | ((z >>> 10) & 127);
            uint8_1[this.x++] = z >>> 17;
        }
        else if (z <= 2147483647) {
            this.ensureCapacity(3);
            var uint8_2 = this.uint8;
            uint8_2[this.x++] = 128 | ((z >>> 10) & 127);
            uint8_2[this.x++] = 128 | ((z >>> 17) & 127);
            uint8_2[this.x++] = z >>> 24;
        }
        else {
            var z1 = z | 0;
            if (z1 < 0)
                z1 += 4294967296;
            var z2 = (z - z1) / 4294967296;
            this.ensureCapacity(4);
            var uint8_3 = this.uint8;
            uint8_3[this.x++] = 128 | ((z1 >>> 10) & 127);
            uint8_3[this.x++] = 128 | ((z1 >>> 17) & 127);
            uint8_3[this.x++] = 128 | ((z1 >>> 24) & 127);
            uint8_3[this.x++] = (z1 >>> 31) | (z2 << 1);
        }
    };
    return CrdtWriter;
}(Writer_1.Writer));
exports.CrdtWriter = CrdtWriter;
