// json-joy/lib/reactive-rpc/common/util/TimedQueue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimedQueue = void 0;
var TimedQueue = (function () {
    function TimedQueue() {
        this.itemLimit = 100;
        this.timeLimit = 5;
        this.onFlush = function (list) { };
        this.list = [];
        this.timer = null;
    }
    TimedQueue.prototype.push = function (item) {
        var _this = this;
        this.list.push(item);
        if (this.list.length >= this.itemLimit) {
            this.flush();
            return;
        }
        if (!this.timer) {
            this.timer = setTimeout(function () {
                _this.flush();
            }, this.timeLimit);
        }
    };
    TimedQueue.prototype.flush = function () {
        var list = this.list;
        this.list = [];
        if (this.timer)
            clearTimeout(this.timer);
        this.timer = null;
        if (list.length)
            this.onFlush(list);
        return list;
    };
    return TimedQueue;
}());
exports.TimedQueue = TimedQueue;
