// json-joy/lib/util/Fuzzer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fuzzer = void 0;
var crypto_1 = require("crypto");
function xoshiro128ss(a, b, c, d) {
    return function () {
        var t = b << 9;
        var r = b * 5;
        r = ((r << 7) | (r >>> 25)) * 9;
        c ^= a;
        d ^= b;
        b ^= c;
        a ^= d;
        c ^= t;
        d = (d << 11) | (d >>> 21);
        return (r >>> 0) / 4294967296;
    };
}
var Fuzzer = (function () {
    function Fuzzer(seed) {
        this.randomInt = function (min, max) {
            return Math.floor(Math.random() * (max - min + 1)) + min;
        };
        this.pick = function (elements) {
            return elements[Math.floor(Math.random() * elements.length)];
        };
        this.repeat = function (times, callback) {
            var result = [];
            for (var i = 0; i < times; i++)
                result.push(callback());
            return result;
        };
        this.seed = seed = seed || (0, crypto_1.randomBytes)(4 * 4);
        var i = 0;
        var a = (seed[i++] << 24) | (seed[i++] << 16) | (seed[i++] << 8) | seed[i++];
        var b = (seed[i++] << 24) | (seed[i++] << 16) | (seed[i++] << 8) | seed[i++];
        var c = (seed[i++] << 24) | (seed[i++] << 16) | (seed[i++] << 8) | seed[i++];
        var d = (seed[i++] << 24) | (seed[i++] << 16) | (seed[i++] << 8) | seed[i++];
        this.random = xoshiro128ss(a, b, c, d);
        Math.random = this.random;
    }
    Fuzzer.randomInt = function (min, max) {
        return Math.floor(Math.random() * (max - min + 1)) + min;
    };
    Fuzzer.pick = function (elements) {
        return elements[Math.floor(Math.random() * elements.length)];
    };
    Fuzzer.repeat = function (times, callback) {
        var result = [];
        for (var i = 0; i < times; i++)
            result.push(callback());
        return result;
    };
    return Fuzzer;
}());
exports.Fuzzer = Fuzzer;
