// json-joy/lib/json-crdt-patch/operations/ConstOp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstOp = void 0;
var clock_1 = require("../clock");
var ConstOp = (function () {
    function ConstOp(id, val) {
        this.id = id;
        this.val = val;
    }
    ConstOp.prototype.span = function () {
        return 1;
    };
    ConstOp.prototype.name = function () {
        return 'const';
    };
    ConstOp.prototype.toString = function (tab) {
        if (tab === void 0) { tab = ''; }
        var val = this.val;
        var valFormatted = val instanceof clock_1.Timestamp
            ? "{ ".concat((0, clock_1.toDisplayString)(val), " }")
            : val instanceof Uint8Array
                ? val.length < 13
                    ? "Uint8Array { ".concat(('' + val).replaceAll(',', ', '), " }")
                    : "Uint8Array(".concat(val.length, ")")
                : "{ ".concat(JSON.stringify(val), " }");
        return "\"".concat(this.name(), "\" ").concat((0, clock_1.toDisplayString)(this.id), " ").concat(valFormatted);
    };
    return ConstOp;
}());
exports.ConstOp = ConstOp;
