// json-joy/lib/json-crdt-patch/operations/ObjSetOp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjSetOp = void 0;
var clock_1 = require("../clock");
var ObjSetOp = (function () {
    function ObjSetOp(id, obj, data) {
        this.id = id;
        this.obj = obj;
        this.data = data;
    }
    ObjSetOp.prototype.span = function () {
        return 0;
    };
    ObjSetOp.prototype.name = function () {
        return 'obj_set';
    };
    ObjSetOp.prototype.toString = function (tab) {
        if (tab === void 0) { tab = ''; }
        var out = "\"".concat(this.name(), "\" ").concat((0, clock_1.toDisplayString)(this.id), "!").concat(this.span(), ", obj = ").concat((0, clock_1.toDisplayString)(this.obj));
        for (var i = 0; i < this.data.length; i++) {
            var isLast = i === this.data.length - 1;
            out += "\n".concat(tab, "  ").concat(isLast ? '└─' : '├─', " ").concat(JSON.stringify(this.data[i][0]), ": ").concat((0, clock_1.toDisplayString)(this.data[i][1]));
        }
        return out;
    };
    return ObjSetOp;
}());
exports.ObjSetOp = ObjSetOp;
