// json-joy/lib/json-patch/op/OpContains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpContains = void 0;
var tslib_1 = require("tslib");
var AbstractPredicateOp_1 = require("./AbstractPredicateOp");
var json_pointer_1 = require("../../json-pointer");
var OpContains = (function (_super) {
    tslib_1.__extends(OpContains, _super);
    function OpContains(path, value, ignore_case) {
        var _this = _super.call(this, path) || this;
        _this.value = value;
        _this.ignore_case = ignore_case;
        return _this;
    }
    OpContains.prototype.op = function () {
        return 'contains';
    };
    OpContains.prototype.code = function () {
        return 30;
    };
    OpContains.prototype.test = function (doc) {
        var val = (0, json_pointer_1.find)(doc, this.path).val;
        if (typeof val !== 'string')
            return false;
        var testValue = this.ignore_case ? val.toLowerCase() : val;
        var testString = this.ignore_case ? this.value.toLowerCase() : this.value;
        var test = testValue.indexOf(testString) > -1;
        return test;
    };
    OpContains.prototype.toJson = function (parent) {
        var op = {
            op: 'contains',
            path: (0, json_pointer_1.formatJsonPointer)(parent ? this.path.slice(parent.path.length) : this.path),
            value: this.value,
        };
        if (this.ignore_case)
            op.ignore_case = this.ignore_case;
        return op;
    };
    OpContains.prototype.toCompact = function (parent, verbose) {
        var opcode = verbose ? 'contains' : 30;
        return this.ignore_case
            ? [opcode, parent ? this.path.slice(parent.path.length) : this.path, this.value, 1]
            : [opcode, parent ? this.path.slice(parent.path.length) : this.path, this.value];
    };
    OpContains.prototype.encode = function (encoder, parent) {
        var ignoreCase = this.ignore_case;
        encoder.encodeArrayHeader(ignoreCase ? 4 : 3);
        encoder.writer.u8(30);
        encoder.encodeArray(parent ? this.path.slice(parent.path.length) : this.path);
        encoder.encodeString(this.value);
        if (ignoreCase)
            encoder.writer.u8(1);
    };
    return OpContains;
}(AbstractPredicateOp_1.AbstractPredicateOp));
exports.OpContains = OpContains;
