// json-joy/lib/json-patch/op/OpDefined.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpDefined = void 0;
var tslib_1 = require("tslib");
var AbstractPredicateOp_1 = require("./AbstractPredicateOp");
var json_pointer_1 = require("../../json-pointer");
var OpDefined = (function (_super) {
    tslib_1.__extends(OpDefined, _super);
    function OpDefined(path) {
        return _super.call(this, path) || this;
    }
    OpDefined.prototype.op = function () {
        return 'defined';
    };
    OpDefined.prototype.code = function () {
        return 31;
    };
    OpDefined.prototype.test = function (doc) {
        var val = (0, json_pointer_1.find)(doc, this.path).val;
        var test = val !== undefined;
        return test;
    };
    OpDefined.prototype.toJson = function (parent) {
        var op = {
            op: 'defined',
            path: (0, json_pointer_1.formatJsonPointer)(parent ? this.path.slice(parent.path.length) : this.path),
        };
        return op;
    };
    OpDefined.prototype.toCompact = function (parent, verbose) {
        var opcode = verbose ? 'defined' : 31;
        return [opcode, parent ? this.path.slice(parent.path.length) : this.path];
    };
    OpDefined.prototype.encode = function (encoder, parent) {
        encoder.encodeArrayHeader(2);
        encoder.writer.u8(31);
        encoder.encodeArray(parent ? this.path.slice(parent.path.length) : this.path);
    };
    return OpDefined;
}(AbstractPredicateOp_1.AbstractPredicateOp));
exports.OpDefined = OpDefined;
