// json-joy/lib/json-patch/op/OpFlip.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpFlip = void 0;
var tslib_1 = require("tslib");
var AbstractOp_1 = require("./AbstractOp");
var json_pointer_1 = require("../../json-pointer");
var OpFlip = (function (_super) {
    tslib_1.__extends(OpFlip, _super);
    function OpFlip(path) {
        return _super.call(this, path) || this;
    }
    OpFlip.prototype.op = function () {
        return 'flip';
    };
    OpFlip.prototype.code = function () {
        return 8;
    };
    OpFlip.prototype.apply = function (doc) {
        var ref = (0, json_pointer_1.find)(doc, this.path);
        if (ref.obj)
            ref.obj[ref.key] = !ref.val;
        else
            doc = !ref.val;
        return { doc: doc, old: ref.val };
    };
    OpFlip.prototype.toJson = function (parent) {
        var op = {
            op: 'flip',
            path: (0, json_pointer_1.formatJsonPointer)(this.path),
        };
        return op;
    };
    OpFlip.prototype.toCompact = function (parent, verbose) {
        var opcode = verbose ? 'flip' : 8;
        return [opcode, this.path];
    };
    OpFlip.prototype.encode = function (encoder, parent) {
        encoder.encodeArrayHeader(2);
        encoder.writer.u8(8);
        encoder.encodeArray(this.path);
    };
    return OpFlip;
}(AbstractOp_1.AbstractOp));
exports.OpFlip = OpFlip;
