// json-joy/lib/json-patch/op/OpInc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpInc = void 0;
var tslib_1 = require("tslib");
var AbstractOp_1 = require("./AbstractOp");
var json_pointer_1 = require("../../json-pointer");
var OpInc = (function (_super) {
    tslib_1.__extends(OpInc, _super);
    function OpInc(path, inc) {
        var _this = _super.call(this, path) || this;
        _this.inc = inc;
        return _this;
    }
    OpInc.prototype.op = function () {
        return 'inc';
    };
    OpInc.prototype.code = function () {
        return 9;
    };
    OpInc.prototype.apply = function (doc) {
        var ref = (0, json_pointer_1.find)(doc, this.path);
        var result = this.inc + Number(ref.val);
        if (ref.obj)
            ref.obj[ref.key] = result;
        else
            doc = result;
        return { doc: doc, old: ref.val };
    };
    OpInc.prototype.toJson = function (parent) {
        var op = {
            op: 'inc',
            path: (0, json_pointer_1.formatJsonPointer)(this.path),
            inc: this.inc,
        };
        return op;
    };
    OpInc.prototype.toCompact = function (parent, verbose) {
        var opcode = verbose ? 'inc' : 9;
        return [opcode, this.path, this.inc];
    };
    OpInc.prototype.encode = function (encoder, parent) {
        encoder.encodeArrayHeader(3);
        encoder.writer.u8(9);
        encoder.encodeArray(this.path);
        encoder.encodeNumber(this.inc);
    };
    return OpInc;
}(AbstractOp_1.AbstractOp));
exports.OpInc = OpInc;
