// json-joy/lib/json-patch/op/OpMatches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpMatches = void 0;
var tslib_1 = require("tslib");
var AbstractPredicateOp_1 = require("./AbstractPredicateOp");
var json_pointer_1 = require("../../json-pointer");
var OpMatches = (function (_super) {
    tslib_1.__extends(OpMatches, _super);
    function OpMatches(path, value, ignore_case, createMatcher) {
        var _this = _super.call(this, path) || this;
        _this.value = value;
        _this.ignore_case = ignore_case;
        _this.matcher = createMatcher(value, ignore_case);
        return _this;
    }
    OpMatches.prototype.op = function () {
        return 'matches';
    };
    OpMatches.prototype.code = function () {
        return 35;
    };
    OpMatches.prototype.test = function (doc) {
        var val = (0, json_pointer_1.find)(doc, this.path).val;
        if (typeof val !== 'string')
            return false;
        var test = this.matcher(val);
        return test;
    };
    OpMatches.prototype.toJson = function (parent) {
        var op = {
            op: 'matches',
            path: (0, json_pointer_1.formatJsonPointer)(parent ? this.path.slice(parent.path.length) : this.path),
            value: this.value,
        };
        if (this.ignore_case)
            op.ignore_case = this.ignore_case;
        return op;
    };
    OpMatches.prototype.toCompact = function (parent, verbose) {
        var opcode = verbose ? 'matches' : 35;
        return this.ignore_case
            ? [opcode, parent ? this.path.slice(parent.path.length) : this.path, this.value, 1]
            : [opcode, parent ? this.path.slice(parent.path.length) : this.path, this.value];
    };
    OpMatches.prototype.encode = function (encoder, parent) {
        var ignoreCase = this.ignore_case;
        encoder.encodeArrayHeader(ignoreCase ? 4 : 3);
        encoder.writer.u8(35);
        encoder.encodeArray(parent ? this.path.slice(parent.path.length) : this.path);
        encoder.encodeString(this.value);
        if (ignoreCase)
            encoder.writer.u8(1);
    };
    return OpMatches;
}(AbstractPredicateOp_1.AbstractPredicateOp));
exports.OpMatches = OpMatches;
