// json-joy/lib/json-patch/op/OpMore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpMore = void 0;
var tslib_1 = require("tslib");
var AbstractPredicateOp_1 = require("./AbstractPredicateOp");
var json_pointer_1 = require("../../json-pointer");
var OpMore = (function (_super) {
    tslib_1.__extends(OpMore, _super);
    function OpMore(path, value) {
        var _this = _super.call(this, path) || this;
        _this.value = value;
        return _this;
    }
    OpMore.prototype.op = function () {
        return 'more';
    };
    OpMore.prototype.code = function () {
        return 36;
    };
    OpMore.prototype.test = function (doc) {
        var val = (0, json_pointer_1.find)(doc, this.path).val;
        if (typeof val !== 'number')
            return false;
        var test = val > this.value;
        return test;
    };
    OpMore.prototype.toJson = function (parent) {
        var op = {
            op: 'more',
            path: (0, json_pointer_1.formatJsonPointer)(parent ? this.path.slice(parent.path.length) : this.path),
            value: this.value,
        };
        return op;
    };
    OpMore.prototype.toCompact = function (parent, verbose) {
        var opcode = verbose ? 'more' : 36;
        return [opcode, parent ? this.path.slice(parent.path.length) : this.path, this.value];
    };
    OpMore.prototype.encode = function (encoder, parent) {
        encoder.encodeArrayHeader(3);
        encoder.writer.u8(36);
        encoder.encodeArray(parent ? this.path.slice(parent.path.length) : this.path);
        encoder.encodeNumber(this.value);
    };
    return OpMore;
}(AbstractPredicateOp_1.AbstractPredicateOp));
exports.OpMore = OpMore;
