// json-joy/lib/json-patch/op/OpMove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpMove = void 0;
var tslib_1 = require("tslib");
var AbstractOp_1 = require("./AbstractOp");
var OpRemove_1 = require("./OpRemove");
var OpAdd_1 = require("./OpAdd");
var json_pointer_1 = require("../../json-pointer");
var OpMove = (function (_super) {
    tslib_1.__extends(OpMove, _super);
    function OpMove(path, from) {
        var _this = _super.call(this, path) || this;
        _this.from = from;
        return _this;
    }
    OpMove.prototype.op = function () {
        return 'move';
    };
    OpMove.prototype.code = function () {
        return 4;
    };
    OpMove.prototype.apply = function (doc) {
        var remove = new OpRemove_1.OpRemove((0, json_pointer_1.toPath)(this.from), undefined).apply(doc);
        var add = new OpAdd_1.OpAdd(this.path, remove.old).apply(remove.doc);
        return add;
    };
    OpMove.prototype.toJson = function (parent) {
        return {
            op: 'move',
            path: (0, json_pointer_1.formatJsonPointer)(this.path),
            from: (0, json_pointer_1.formatJsonPointer)(this.from),
        };
    };
    OpMove.prototype.toCompact = function (parent, verbose) {
        var opcode = verbose ? 'move' : 4;
        return [opcode, this.path, this.from];
    };
    OpMove.prototype.encode = function (encoder, parent) {
        encoder.encodeArrayHeader(3);
        encoder.writer.u8(4);
        encoder.encodeArray(this.path);
        encoder.encodeArray(this.from);
    };
    return OpMove;
}(AbstractOp_1.AbstractOp));
exports.OpMove = OpMove;
