// json-joy/lib/json-patch/op/OpReplace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpReplace = void 0;
var tslib_1 = require("tslib");
var AbstractOp_1 = require("./AbstractOp");
var json_pointer_1 = require("../../json-pointer");
var OpReplace = (function (_super) {
    tslib_1.__extends(OpReplace, _super);
    function OpReplace(path, value, oldValue) {
        var _this = _super.call(this, path) || this;
        _this.value = value;
        _this.oldValue = oldValue;
        return _this;
    }
    OpReplace.prototype.op = function () {
        return 'replace';
    };
    OpReplace.prototype.code = function () {
        return 2;
    };
    OpReplace.prototype.apply = function (doc) {
        var ref = (0, json_pointer_1.find)(doc, this.path);
        if (ref.val === undefined)
            throw new Error('NOT_FOUND');
        if ((0, json_pointer_1.isObjectReference)(ref))
            ref.obj[ref.key] = this.value;
        else if ((0, json_pointer_1.isArrayReference)(ref))
            ref.obj[ref.key] = this.value;
        else
            doc = this.value;
        return { doc: doc, old: ref.val };
    };
    OpReplace.prototype.toJson = function (parent) {
        var json = {
            op: 'replace',
            path: (0, json_pointer_1.formatJsonPointer)(this.path),
            value: this.value,
        };
        if (this.oldValue !== undefined)
            json.oldValue = this.oldValue;
        return json;
    };
    OpReplace.prototype.toCompact = function (parent, verbose) {
        var opcode = verbose ? 'replace' : 2;
        return this.oldValue === undefined
            ? [opcode, this.path, this.value]
            : [opcode, this.path, this.value, this.oldValue];
    };
    OpReplace.prototype.encode = function (encoder, parent) {
        var hasOldValue = this.oldValue !== undefined;
        encoder.encodeArrayHeader(hasOldValue ? 4 : 3);
        encoder.writer.u8(2);
        encoder.encodeArray(this.path);
        encoder.encodeAny(this.value);
        if (hasOldValue)
            encoder.encodeAny(this.oldValue);
    };
    return OpReplace;
}(AbstractOp_1.AbstractOp));
exports.OpReplace = OpReplace;
