// json-joy/lib/json-patch/op/OpTestString.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpTestString = void 0;
var tslib_1 = require("tslib");
var AbstractPredicateOp_1 = require("./AbstractPredicateOp");
var json_pointer_1 = require("../../json-pointer");
var OpTestString = (function (_super) {
    tslib_1.__extends(OpTestString, _super);
    function OpTestString(path, pos, str, not) {
        var _this = _super.call(this, path) || this;
        _this.pos = pos;
        _this.str = str;
        _this.not = not;
        return _this;
    }
    OpTestString.prototype.op = function () {
        return 'test_string';
    };
    OpTestString.prototype.code = function () {
        return 40;
    };
    OpTestString.prototype.test = function (doc) {
        var val = (0, json_pointer_1.find)(doc, this.path).val;
        if (typeof val !== 'string')
            return false;
        var length = val.length;
        var start = Math.min(this.pos, length);
        var end = Math.min(this.pos + this.str.length, length);
        var test = val.substring(start, end) === this.str;
        return this.not ? !test : test;
    };
    OpTestString.prototype.toJson = function (parent) {
        var op = {
            op: 'test_string',
            path: (0, json_pointer_1.formatJsonPointer)(parent ? this.path.slice(parent.path.length) : this.path),
            pos: this.pos,
            str: this.str,
        };
        if (this.not)
            op.not = this.not;
        return op;
    };
    OpTestString.prototype.toCompact = function (parent, verbose) {
        var opcode = verbose ? 'test_string' : 40;
        var path = parent ? this.path.slice(parent.path.length) : this.path;
        return this.not ? [opcode, path, this.pos, this.str, 1] : [opcode, path, this.pos, this.str];
    };
    OpTestString.prototype.encode = function (encoder, parent) {
        encoder.encodeArrayHeader(this.not ? 5 : 4);
        encoder.writer.u8(40);
        encoder.encodeArray(parent ? this.path.slice(parent.path.length) : this.path);
        encoder.encodeNumber(this.pos);
        encoder.encodeString(this.str);
        if (this.not)
            encoder.writer.u8(1);
    };
    return OpTestString;
}(AbstractPredicateOp_1.AbstractPredicateOp));
exports.OpTestString = OpTestString;
