// json-joy/lib/json-patch/op/OpType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpType = void 0;
var tslib_1 = require("tslib");
var AbstractPredicateOp_1 = require("./AbstractPredicateOp");
var json_pointer_1 = require("../../json-pointer");
var isArray = Array.isArray;
var OpType = (function (_super) {
    tslib_1.__extends(OpType, _super);
    function OpType(path, value) {
        var _this = _super.call(this, path) || this;
        _this.value = value;
        return _this;
    }
    OpType.prototype.op = function () {
        return 'type';
    };
    OpType.prototype.code = function () {
        return 42;
    };
    OpType.prototype.test = function (doc) {
        var val = (0, json_pointer_1.find)(doc, this.path).val;
        if (val === null)
            return this.value === 'null';
        if (isArray(val))
            return this.value === 'array';
        if (typeof val === this.value)
            return true;
        if (typeof val === 'number' && val === Math.round(val) && this.value === 'integer')
            return true;
        return false;
    };
    OpType.prototype.toJson = function (parent) {
        var op = {
            op: 'type',
            path: (0, json_pointer_1.formatJsonPointer)(parent ? this.path.slice(parent.path.length) : this.path),
            value: this.value,
        };
        return op;
    };
    OpType.prototype.toCompact = function (parent, verbose) {
        var opcode = verbose ? 'type' : 42;
        return [opcode, parent ? this.path.slice(parent.path.length) : this.path, this.value];
    };
    OpType.prototype.encode = function (encoder, parent) {
        encoder.encodeArrayHeader(3);
        encoder.writer.u8(42);
        encoder.encodeArray(parent ? this.path.slice(parent.path.length) : this.path);
        encoder.encodeString(this.value);
    };
    return OpType;
}(AbstractPredicateOp_1.AbstractPredicateOp));
exports.OpType = OpType;
