// json-joy/lib/util/trees/trie/TrieNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrieNode = void 0;
var radix_1 = require("../radix/radix");
var util_1 = require("../util");
var TrieNode = (function () {
    function TrieNode(k, v) {
        this.k = k;
        this.v = v;
        this.p = undefined;
        this.l = undefined;
        this.r = undefined;
        this.children = undefined;
    }
    TrieNode.prototype.forChildren = function (callback) {
        var child = (0, util_1.first)(this.children);
        var i = 0;
        while (child) {
            callback(child, 0);
            i++;
            child = (0, util_1.next)(child);
        }
    };
    TrieNode.prototype.toRecord = function (prefix, record) {
        return (0, radix_1.toRecord)(this, prefix, record);
    };
    TrieNode.prototype.toString = function (tab) {
        if (tab === void 0) { tab = ''; }
        return (0, radix_1.print)(this, tab);
    };
    return TrieNode;
}());
exports.TrieNode = TrieNode;
