// json-joy/lib/json-pack/msgpack/MsgPackEncoderFast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgPackEncoderFast = void 0;
var Writer_1 = require("../../util/buffers/Writer");
var MsgPackEncoderFast = (function () {
    function MsgPackEncoderFast(writer) {
        if (writer === void 0) { writer = new Writer_1.Writer(); }
        this.writer = writer;
    }
    MsgPackEncoderFast.prototype.encode = function (json) {
        this.writer.reset();
        this.writeAny(json);
        return this.writer.flush();
    };
    MsgPackEncoderFast.prototype.encodeAny = function (json) {
        this.writeAny(json);
    };
    MsgPackEncoderFast.prototype.writeAny = function (value) {
        switch (value) {
            case null:
                return this.writer.u8(0xc0);
            case false:
                return this.writer.u8(0xc2);
            case true:
                return this.writer.u8(0xc3);
        }
        if (value instanceof Array)
            return this.writeArr(value);
        switch (typeof value) {
            case 'number':
                return this.writeNumber(value);
            case 'string':
                return this.writeStr(value);
            case 'object':
                return this.writeObj(value);
        }
    };
    MsgPackEncoderFast.prototype.encodeFloat64 = function (num) {
        this.writeFloat(num);
    };
    MsgPackEncoderFast.prototype.writeNull = function () {
        return this.writer.u8(0xc0);
    };
    MsgPackEncoderFast.prototype.writeFloat = function (float) {
        this.writer.u8f64(0xcb, float);
    };
    MsgPackEncoderFast.prototype.u32 = function (num) {
        var writer = this.writer;
        this.writer.ensureCapacity(5);
        var uint8 = writer.uint8;
        if (num <= 127) {
            uint8[writer.x++] = num;
        }
        else if (num <= 0xffff) {
            uint8[writer.x++] = 0xcd;
            writer.view.setUint16(writer.x, num);
            writer.x += 2;
        }
        else if (num <= 0xffffffff) {
            uint8[writer.x++] = 0xce;
            writer.view.setUint32(writer.x, num);
            writer.x += 4;
        }
        else
            this.writeFloat(num);
    };
    MsgPackEncoderFast.prototype.n32 = function (num) {
        var writer = this.writer;
        this.writer.ensureCapacity(5);
        var uint8 = writer.uint8;
        if (num >= -0x20) {
            uint8[writer.x++] = 0x100 + num;
        }
        else if (num >= -0x8000) {
            uint8[writer.x++] = 0xd1;
            writer.view.setInt16(writer.x, num);
            writer.x += 2;
        }
        else if (num >= -0x80000000) {
            uint8[writer.x++] = 0xd2;
            writer.view.setInt32(writer.x, num);
            writer.x += 4;
        }
        else
            this.writeFloat(num);
    };
    MsgPackEncoderFast.prototype.encodeNumber = function (num) {
        this.writeNumber(num);
    };
    MsgPackEncoderFast.prototype.writeNumber = function (num) {
        if (num >>> 0 === num)
            return this.u32(num);
        if (num >> 0 === num)
            return this.n32(num);
        this.writeFloat(num);
    };
    MsgPackEncoderFast.prototype.writeInteger = function (int) {
        if (int >= 0)
            if (int <= 0xffffffff)
                return this.u32(int);
            else if (int > -0x80000000)
                return this.n32(int);
        this.writeFloat(int);
    };
    MsgPackEncoderFast.prototype.writeUInteger = function (uint) {
        if (uint <= 0xffffffff)
            return this.u32(uint);
        this.writeFloat(uint);
    };
    MsgPackEncoderFast.prototype.encodeNull = function () {
        this.writer.u8(0xc0);
    };
    MsgPackEncoderFast.prototype.encodeTrue = function () {
        this.writer.u8(0xc3);
    };
    MsgPackEncoderFast.prototype.encodeFalse = function () {
        this.writer.u8(0xc2);
    };
    MsgPackEncoderFast.prototype.encodeBoolean = function (bool) {
        this.writeBoolean(bool);
    };
    MsgPackEncoderFast.prototype.writeBoolean = function (bool) {
        if (bool)
            this.writer.u8(0xc3);
        else
            this.writer.u8(0xc2);
    };
    MsgPackEncoderFast.prototype.encodeStringHeader = function (length) {
        this.writeStrHdr(length);
    };
    MsgPackEncoderFast.prototype.writeStrHdr = function (length) {
        if (length <= 31)
            this.writer.u8(160 | length);
        else if (length <= 0xff)
            this.writer.u16(0xd900 + length);
        else if (length <= 0xffff)
            this.writer.u8u16(0xda, length);
        else
            this.writer.u8u32(0xdb, length);
    };
    MsgPackEncoderFast.prototype.encodeString = function (str) {
        this.writeStr(str);
    };
    MsgPackEncoderFast.prototype.writeStr = function (str) {
        var writer = this.writer;
        var length = str.length;
        var maxSize = length * 4;
        writer.ensureCapacity(5 + maxSize);
        var uint8 = writer.uint8;
        var lengthOffset = writer.x;
        if (maxSize <= 31)
            writer.x++;
        else if (maxSize <= 0xff) {
            uint8[writer.x++] = 0xd9;
            lengthOffset = writer.x;
            writer.x++;
        }
        else if (maxSize <= 0xffff) {
            uint8[writer.x++] = 0xda;
            lengthOffset = writer.x;
            writer.x += 2;
        }
        else {
            uint8[writer.x++] = 0xdb;
            lengthOffset = writer.x;
            writer.x += 4;
        }
        var bytesWritten = this.writer.utf8(str);
        if (maxSize <= 31)
            uint8[lengthOffset] = 160 | bytesWritten;
        else if (maxSize <= 0xff)
            uint8[lengthOffset] = bytesWritten;
        else if (maxSize <= 0xffff)
            writer.view.setUint16(lengthOffset, bytesWritten);
        else
            writer.view.setUint32(lengthOffset, bytesWritten);
    };
    MsgPackEncoderFast.prototype.encodeAsciiString = function (str) {
        this.writeAsciiStr(str);
    };
    MsgPackEncoderFast.prototype.writeAsciiStr = function (str) {
        this.writeStrHdr(str.length);
        this.writer.ascii(str);
    };
    MsgPackEncoderFast.prototype.encodeArrayHeader = function (length) {
        this.writeArrHdr(length);
    };
    MsgPackEncoderFast.prototype.encodeArray = function (arr) {
        this.writeArr(arr);
    };
    MsgPackEncoderFast.prototype.writeArrHdr = function (length) {
        if (length <= 15)
            this.writer.u8(144 | length);
        else if (length <= 0xffff)
            this.writer.u8u16(0xdc, length);
        else if (length <= 0xffffffff)
            this.writer.u8u32(0xdd, length);
    };
    MsgPackEncoderFast.prototype.writeArr = function (arr) {
        var length = arr.length;
        if (length <= 15)
            this.writer.u8(144 | length);
        else if (length <= 0xffff)
            this.writer.u8u16(0xdc, length);
        else if (length <= 0xffffffff)
            this.writer.u8u32(0xdd, length);
        for (var i = 0; i < length; i++)
            this.writeAny(arr[i]);
    };
    MsgPackEncoderFast.prototype.encodeObjectHeader = function (length) {
        this.writeObjHdr(length);
    };
    MsgPackEncoderFast.prototype.encodeObject = function (obj) {
        this.writeObj(obj);
    };
    MsgPackEncoderFast.prototype.writeObjHdr = function (length) {
        if (length <= 15)
            this.writer.u8(128 | length);
        else if (length <= 0xffff) {
            this.writer.u8u16(0xde, length);
        }
        else if (length <= 0xffffffff) {
            this.writer.u8u32(0xdf, length);
        }
    };
    MsgPackEncoderFast.prototype.writeObj = function (obj) {
        var keys = Object.keys(obj);
        var length = keys.length;
        this.writeObjHdr(length);
        for (var i = 0; i < length; i++) {
            var key = keys[i];
            this.writeStr(key);
            this.writeAny(obj[key]);
        }
    };
    MsgPackEncoderFast.prototype.encodeExtHeader = function (type, length) {
        switch (length) {
            case 1:
                this.writer.u16((0xd4 << 8) | type);
                break;
            case 2:
                this.writer.u16((0xd5 << 8) | type);
                break;
            case 4:
                this.writer.u16((0xd6 << 8) | type);
                break;
            case 8:
                this.writer.u16((0xd7 << 8) | type);
                break;
            case 16:
                this.writer.u16((0xd8 << 8) | type);
                break;
            default:
                if (length <= 0xff) {
                    this.writer.u16((0xc7 << 8) | length);
                    this.writer.u8(type);
                }
                else if (length <= 0xffff) {
                    this.writer.u8u16(0xc8, length);
                    this.writer.u8(type);
                }
                else if (length <= 0xffffffff) {
                    this.writer.u8u32(0xc9, length);
                    this.writer.u8(type);
                }
        }
    };
    MsgPackEncoderFast.prototype.encodeExt = function (ext) {
        var type = ext.tag, buf = ext.val;
        var length = buf.length;
        this.encodeExtHeader(type, length);
        this.writer.buf(buf, length);
    };
    MsgPackEncoderFast.prototype.encodeBinaryHeader = function (length) {
        this.writeBinHdr(length);
    };
    MsgPackEncoderFast.prototype.encodeBinary = function (buf) {
        this.writeBin(buf);
    };
    MsgPackEncoderFast.prototype.writeBinHdr = function (length) {
        if (length <= 0xff)
            this.writer.u16((0xc4 << 8) | length);
        else if (length <= 0xffff) {
            this.writer.u8u16(0xc5, length);
        }
        else if (length <= 0xffffffff) {
            this.writer.u8u32(0xc6, length);
        }
    };
    MsgPackEncoderFast.prototype.writeBin = function (buf) {
        var length = buf.length;
        this.writeBinHdr(length);
        this.writer.buf(buf, length);
    };
    return MsgPackEncoderFast;
}());
exports.MsgPackEncoderFast = MsgPackEncoderFast;
