// json-joy/lib/json-crdt/types/const/Const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Const = void 0;
var clock_1 = require("../../../json-crdt-patch/clock");
var Const = (function () {
    function Const(id, val) {
        this.id = id;
        this.val = val;
    }
    Const.prototype.view = function () {
        return this.val;
    };
    Const.prototype.children = function () { };
    Const.prototype.child = function () {
        return undefined;
    };
    Const.prototype.container = function () {
        return undefined;
    };
    Const.prototype.toString = function (tab) {
        var val = this.val;
        var valFormatted = val instanceof Uint8Array
            ? "Uint8Array { ".concat(('' + val).replaceAll(',', ', '), " }")
            : "{ ".concat(val instanceof clock_1.Timestamp ? (0, clock_1.toDisplayString)(val) : JSON.stringify(val), " }");
        return "".concat(this.constructor.name, " ").concat((0, clock_1.toDisplayString)(this.id), " ").concat(valFormatted);
    };
    return Const;
}());
exports.Const = Const;
